<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Accessing EuPathDB Resources using AnnotationHub</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>

</head>

<body>




<h1 class="title toc-ignore">Accessing EuPathDB Resources using AnnotationHub</h1>



<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<p><strong>Authors</strong>: <a href="mailto:keith.hughitt@nih.gov">V. Keith Hughitt</a><br /> <strong>Modified:</strong> 2019-02-12 16:56:30<br /> <strong>Compiled</strong>: Wed Feb 27 10:29:56 2019</p>
<div id="overview" class="section level1">
<h1>Overview</h1>
<p>This tutorial describes how to query and make use of annotations retrieved from <a href="http://eupathdb.org/eupathdb/">EuPathDB : The Eukaryotic Pathogen Genomics Resource</a> using <a href="http://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a>.</p>
<p>For more information on using AnnotationHub, check out the AnnotationHub vignettes:</p>
<ul>
<li><a href="http://bioconductor.org/packages/release/bioc/vignettes/AnnotationHub/inst/doc/AnnotationHub-HOWTO.html">AnnotationHub: Access the AnnotationHub Web Service</a></li>
<li><a href="http://bioconductor.org/packages/release/bioc/vignettes/AnnotationHub/inst/doc/AnnotationHub-HOWTO.html">AnnotationHub How-To’s</a></li>
</ul>
<p>The resources described in this tutorial were generating using GFF files and web API requests made to the various EuPathDB databases (TriTrypDB, ToxoDB, etc.) Only organisms with annotated genomes (those for which GFF files are available) are accessible through AnnotationHub.</p>
<p>The two main resources provided are:</p>
<ul>
<li><a href="https://www.bioconductor.org/help/workflows/annotation/annotation/#OrgDb">OrgDb</a></li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GRanges</a></li>
</ul>
<p>OrgDb objects for an organism include basic gene-level information such as:</p>
<ul>
<li>Gene ID</li>
<li>Gene description</li>
<li>Chromosome number</li>
<li>GO terms associated with gene</li>
<li>KEGG Pathways associated with gene</li>
<li>Etc.</li>
</ul>
<p>For some organisms, <a href="https://www.ebi.ac.uk/interpro/">InterPro</a> protein domain information is also available (in some cases, however, even though InterPro domain information is available through EuPathDB, it is too large to be included in the current AnnotationHub resources).</p>
<p>For more information about working with Bioconductor annotation resources, see:</p>
<ul>
<li><a href="https://www.bioconductor.org/help/workflows/annotation/annotation/">Genomic Annotation Resources in Bioconductor</a></li>
</ul>
</div>
<div id="installation" class="section level1">
<h1>Installation</h1>
<p>If you don’t already have AnnotationHub installed on your system, use <code>BiocManager::install</code> to install the package:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</a>
<a class="sourceLine" id="cb1-2" title="2">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;AnnotationHub&quot;</span>)</a></code></pre></div>
</div>
<div id="getting-started" class="section level1">
<h1>Getting started</h1>
<p>To begin, let’s create a new <code>AnnotationHub</code> connection and use it to query AnnotationHub for all EuPathDB resources.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw">library</span>(<span class="st">&quot;EuPathDB&quot;</span>)</a></code></pre></div>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: 'BiocGenerics'</code></pre>
<pre><code>## The following objects are masked from 'package:parallel':
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     anyDuplicated, append, as.data.frame, basename, cbind,
##     colMeans, colnames, colSums, dirname, do.call, duplicated,
##     eval, evalq, Filter, Find, get, grep, grepl, intersect,
##     is.unsorted, lapply, lengths, Map, mapply, match, mget, order,
##     paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind,
##     Reduce, rowMeans, rownames, rowSums, sapply, setdiff, sort,
##     table, tapply, union, unique, unsplit, which, which.max,
##     which.min</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     'browseVignettes()'. To cite Bioconductor, see
##     'citation(&quot;Biobase&quot;)', and for packages 'citation(&quot;pkgname&quot;)'.</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## 
## Attaching package: 'S4Vectors'</code></pre>
<pre><code>## The following object is masked from 'package:base':
## 
##     expand.grid</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: GenomeInfoDbData</code></pre>
<pre><code>## Loading required package: AnnotationHub</code></pre>
<pre><code>## 
## Attaching package: 'AnnotationHub'</code></pre>
<pre><code>## The following object is masked from 'package:Biobase':
## 
##     cache</code></pre>
<pre><code>## 
## This is EuPathDB version 1.0.1
##  Read 'EuPathDB()' to get started.</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" title="1"><span class="kw">library</span>(<span class="st">'AnnotationHub'</span>)</a>
<a class="sourceLine" id="cb23-2" title="2"></a>
<a class="sourceLine" id="cb23-3" title="3"><span class="co"># create an AnnotationHub connection</span></a>
<a class="sourceLine" id="cb23-4" title="4">ah &lt;-<span class="st"> </span><span class="kw">AnnotationHub</span>()</a></code></pre></div>
<pre><code>## snapshotDate(): 2018-10-24</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" title="1"><span class="co"># search for all EuPathDB resources</span></a>
<a class="sourceLine" id="cb25-2" title="2">meta &lt;-<span class="st"> </span><span class="kw">query</span>(ah, <span class="st">&quot;EuPathDB&quot;</span>)</a>
<a class="sourceLine" id="cb25-3" title="3"></a>
<a class="sourceLine" id="cb25-4" title="4"><span class="kw">length</span>(meta)</a></code></pre></div>
<pre><code>## [1] 530</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" title="1"><span class="kw">head</span>(meta)</a></code></pre></div>
<pre><code>## AnnotationHub with 6 records
## # snapshotDate(): 2018-10-24 
## # $dataprovider: FungiDB, TriTrypDB, MicrosporidiaDB
## # $species: Aspergillus sydowii, Aspergillus versicolor, Edhazardia aed...
## # $rdataclass: OrgDb
## # additional mcols(): taxonomyid, genome, description,
## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass,
## #   tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., 'object[[&quot;AH65001&quot;]]' 
## 
##             title                                                         
##   AH65001 | Edhazardia aedis USNM 41457 genome wide annotations           
##   AH65002 | Aspergillus versicolor CBS 583.65 genome wide annotations     
##   AH65003 | Aspergillus sydowii CBS 593.65 genome wide annotations        
##   AH65004 | Phytophthora cinnamomi var. cinnamomi CBS 144.22 genome wid...
##   AH65005 | Trypanosoma vivax Y486 genome wide annotations                
##   AH65006 | Trypanosoma congolense IL3000 genome wide annotations</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" title="1"><span class="co"># types of EuPathDB data available</span></a>
<a class="sourceLine" id="cb29-2" title="2"><span class="kw">table</span>(meta<span class="op">$</span>rdataclass)</a></code></pre></div>
<pre><code>## 
## GRanges   OrgDb 
##     265     265</code></pre>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" title="1"><span class="co"># distribution of resources by specific databases</span></a>
<a class="sourceLine" id="cb31-2" title="2"><span class="kw">table</span>(meta<span class="op">$</span>dataprovider)</a></code></pre></div>
<pre><code>## 
##        AmoebaDB        CryptoDB         FungiDB       GiardiaDB 
##              22              26             228               2 
## MicrosporidiaDB    PiroplasmaDB        PlasmoDB          ToxoDB 
##              60              18              54              52 
##         TrichDB       TriTrypDB 
##               2              66</code></pre>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" title="1"><span class="co"># list of organisms for which resources are available</span></a>
<a class="sourceLine" id="cb33-2" title="2"><span class="kw">length</span>(<span class="kw">unique</span>(meta<span class="op">$</span>species))</a></code></pre></div>
<pre><code>## [1] 196</code></pre>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" title="1"><span class="kw">head</span>(<span class="kw">unique</span>(meta<span class="op">$</span>species))</a></code></pre></div>
<pre><code>## [1] &quot;Edhazardia aedis&quot;       &quot;Aspergillus versicolor&quot;
## [3] &quot;Aspergillus sydowii&quot;    &quot;Phytophthora cinnamomi&quot;
## [5] &quot;Trypanosoma vivax&quot;      &quot;Trypanosoma congolense&quot;</code></pre>
</div>
<div id="working-with-eupathdb-orgdb-resources" class="section level1">
<h1>Working with EuPathDB OrgDb resources</h1>
<p>Next, we will see how you can query AnnotationHub for EuPathDB OrgDb resources.</p>
<p>To begin, create an AnnotationHub connection, if you have not already done so, as shown in the section above.</p>
<p>You can now use the <code>query</code> function to search for your organism of interest and store the result as follows:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" title="1">res &lt;-<span class="st"> </span><span class="kw">query</span>(ah, <span class="kw">c</span>(<span class="st">'Leishmania major strain Friedlin'</span>, <span class="st">'OrgDb'</span>, <span class="st">'EuPathDB'</span>))</a>
<a class="sourceLine" id="cb37-2" title="2">res</a></code></pre></div>
<pre><code>## AnnotationHub with 1 record
## # snapshotDate(): 2018-10-24 
## # names(): AH65089
## # $dataprovider: TriTrypDB
## # $species: Leishmania major
## # $rdataclass: OrgDb
## # $rdatadateadded: 2018-10-15
## # $title: Leishmania major strain Friedlin genome wide annotations
## # $description: TriTrypDB 39 genome annotations for Leishmania major st...
## # $taxonomyid: 347515
## # $genome: TriTrypDB-39_LmajorFriedlin
## # $sourcetype: GFF
## # $sourceurl: http://TriTrypDB.org/common/downloads/Current_Release/Lma...
## # $sourcesize: NA
## # $tags:
## #   c(&quot;Annotation,EuPathDB,Eukaryote,Pathogen,Parasite,Trypanosome,Kinetoplastid,Leishmania&quot;,
## #   &quot;EuPathDB&quot;) 
## # retrieve record with 'object[[&quot;AH65089&quot;]]'</code></pre>
<p>The result includes a single record, “AH56967”. The record can be accessed from the result variable using list-like indexing:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" title="1">orgdb &lt;-<span class="st"> </span>res[[<span class="st">'AH65089'</span>]]</a></code></pre></div>
<pre><code>## downloading 0 resources</code></pre>
<pre><code>## loading from cache 
##     '/home/trey//.AnnotationHub/71835'</code></pre>
<pre><code>## Loading required package: AnnotationDbi</code></pre>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" title="1"><span class="kw">class</span>(orgdb)</a></code></pre></div>
<pre><code>## [1] &quot;OrgDb&quot;
## attr(,&quot;package&quot;)
## [1] &quot;AnnotationDbi&quot;</code></pre>
<p>We can see that we now have an OrgDb instance, and as such, we can use the usual methods available for working this OrgDb objects, including:</p>
<ul>
<li><code>columns()</code></li>
<li><code>keys()</code></li>
<li><code>select()</code></li>
</ul>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" title="1"><span class="co"># list available fields to retrieve</span></a>
<a class="sourceLine" id="cb45-2" title="2"><span class="kw">columns</span>(orgdb)</a></code></pre></div>
<pre><code>##  [1] &quot;CHR&quot;                   &quot;EVIDENCE_CODE&quot;        
##  [3] &quot;GENEDESCRIPTION&quot;       &quot;GENESOURCE&quot;           
##  [5] &quot;GENETYPE&quot;              &quot;GID&quot;                  
##  [7] &quot;GO_ID&quot;                 &quot;GO_TERM_NAME&quot;         
##  [9] &quot;INTERPRO_DESC&quot;         &quot;INTERPRO_E_VALUE&quot;     
## [11] &quot;INTERPRO_END_MIN&quot;      &quot;INTERPRO_FAMILY_ID&quot;   
## [13] &quot;INTERPRO_NAME&quot;         &quot;INTERPRO_PRIMARY_ID&quot;  
## [15] &quot;INTERPRO_SECONDARY_ID&quot; &quot;INTERPRO_START_MIN&quot;   
## [17] &quot;ONTOLOGY&quot;              &quot;PATHWAY&quot;              
## [19] &quot;PATHWAY_SOURCE&quot;        &quot;PATHWAY_SOURCE_ID&quot;    
## [21] &quot;SOURCE&quot;                &quot;TRANSCRIPT_IDS&quot;       
## [23] &quot;TYPE&quot;</code></pre>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" title="1"><span class="co"># create a vector containing all gene ids for the organism</span></a>
<a class="sourceLine" id="cb47-2" title="2">gids &lt;-<span class="st"> </span><span class="kw">keys</span>(orgdb, <span class="dt">keytype=</span><span class="st">'GID'</span>)</a>
<a class="sourceLine" id="cb47-3" title="3"><span class="kw">head</span>(gids)</a></code></pre></div>
<pre><code>## [1] &quot;LmjF.24.0370&quot; &quot;LmjF.34.2180&quot; &quot;LmjF.36.2890&quot; &quot;LmjF.28.0420&quot;
## [5] &quot;LmjF.16.0980&quot; &quot;LmjF.22.0730&quot;</code></pre>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" title="1"><span class="co"># retrieve the chromosome, description, and biotype for each gene</span></a>
<a class="sourceLine" id="cb49-2" title="2">dat &lt;-<span class="st"> </span><span class="kw">select</span>(orgdb, <span class="dt">keys=</span>gids, <span class="dt">keytype=</span><span class="st">'GID'</span>, <span class="dt">columns=</span><span class="kw">c</span>(<span class="st">'CHR'</span>, <span class="st">'TYPE'</span>, <span class="st">'GENEDESCRIPTION'</span>))</a></code></pre></div>
<pre><code>## 'select()' returned 1:1 mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb51-1" title="1"><span class="kw">head</span>(dat)</a></code></pre></div>
<pre><code>##            GID     CHR           TYPE
## 1 LmjF.24.0370 LmjF.24 protein coding
## 2 LmjF.34.2180 LmjF.34 protein coding
## 3 LmjF.36.2890 LmjF.36 protein coding
## 4 LmjF.28.0420 LmjF.28 protein coding
## 5 LmjF.16.0980 LmjF.16 protein coding
## 6 LmjF.22.0730 LmjF.22 protein coding
##                                                GENEDESCRIPTION
## 1                         aspartate aminotransferase, putative
## 2                              hypothetical protein, conserved
## 3 ATP-binding cassette protein subfamily G, member 6, putative
## 4                 hypothetical protein, conserved (pseudogene)
## 5                              hypothetical protein, conserved
## 6                    cytoskeleton associated protein, putative</code></pre>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb53-1" title="1"><span class="kw">table</span>(dat<span class="op">$</span>TYPE)</a></code></pre></div>
<pre><code>## 
## non protein coding     protein coding      rRNA encoding 
##                 83               8402                 63 
##    snoRNA encoding     snRNA encoding      tRNA encoding 
##                741                  6                 83</code></pre>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb55-1" title="1"><span class="kw">table</span>(dat<span class="op">$</span>CHR)</a></code></pre></div>
<pre><code>## 
## LmjF.01 LmjF.02 LmjF.03 LmjF.04 LmjF.05 LmjF.06 LmjF.07 LmjF.08 LmjF.09 
##      85     136      97     130     159     135     130     136     182 
## LmjF.10 LmjF.11 LmjF.12 LmjF.13 LmjF.14 LmjF.15 LmjF.16 LmjF.17 LmjF.18 
##     150     148     142     170     159     192     177     162     219 
## LmjF.19 LmjF.20 LmjF.21 LmjF.22 LmjF.23 LmjF.24 LmjF.25 LmjF.26 LmjF.27 
##     179     276     233     181     229     246     322     495     385 
## LmjF.28 LmjF.29 LmjF.30 LmjF.31 LmjF.32 LmjF.33 LmjF.34 LmjF.35 LmjF.36 
##     319     300     408     346     430     424     500     615     781</code></pre>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb57-1" title="1"><span class="co"># create a gene / GO term mapping</span></a>
<a class="sourceLine" id="cb57-2" title="2">gene_go_mapping &lt;-<span class="st"> </span><span class="kw">select</span>(orgdb, <span class="dt">keys=</span>gids, <span class="dt">keytype=</span><span class="st">'GID'</span>, <span class="dt">columns=</span><span class="kw">c</span>(<span class="st">'GO_ID'</span>, <span class="st">'GO_TERM_NAME'</span>, <span class="st">'ONTOLOGY'</span>))</a></code></pre></div>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb59-1" title="1"><span class="kw">head</span>(gene_go_mapping)</a></code></pre></div>
<pre><code>##            GID      GO_ID                GO_TERM_NAME           ONTOLOGY
## 1 LmjF.24.0370 GO:0009058        biosynthetic process Biological Process
## 2 LmjF.24.0370 GO:0005737                   cytoplasm Cellular Component
## 3 LmjF.24.0370 GO:0031981               nuclear lumen Cellular Component
## 4 LmjF.24.0370 GO:0097014               ciliary plasm Cellular Component
## 5 LmjF.24.0370 GO:0030170 pyridoxal phosphate binding Molecular Function
## 6 LmjF.34.2180 GO:0005488                     binding Molecular Function</code></pre>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb61-1" title="1"><span class="co"># retrieve KEGG, etc. pathway annotations</span></a>
<a class="sourceLine" id="cb61-2" title="2">gene_pathway_mapping &lt;-<span class="st"> </span><span class="kw">select</span>(orgdb, <span class="dt">keys=</span>gids, <span class="dt">keytype=</span><span class="st">'GID'</span>, <span class="dt">columns=</span><span class="kw">c</span>(<span class="st">'PATHWAY'</span>, <span class="st">'PATHWAY_SOURCE'</span>))</a></code></pre></div>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb63-1" title="1"><span class="kw">table</span>(gene_pathway_mapping<span class="op">$</span>PATHWAY_SOURCE)</a></code></pre></div>
<pre><code>## 
##       KEGG   LeishCyc    MetaCyc TrypanoCyc 
##      15803       2367      66680       2475</code></pre>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb65-1" title="1"><span class="kw">head</span>(gene_pathway_mapping)</a></code></pre></div>
<pre><code>##            GID                                                   PATHWAY
## 1 LmjF.24.0370                  L-asparagine degradation III (mammalian)
## 2 LmjF.24.0370                                 L-aspartate degradation I
## 3 LmjF.24.0370                                    aspartate biosynthesis
## 4 LmjF.24.0370                                  L-aspartate biosynthesis
## 5 LmjF.24.0370                                    aspartate biosynthesis
## 6 LmjF.24.0370 superpathway of L-aspartate and L-asparagine biosynthesis
##   PATHWAY_SOURCE
## 1        MetaCyc
## 2        MetaCyc
## 3       LeishCyc
## 4        MetaCyc
## 5     TrypanoCyc
## 6        MetaCyc</code></pre>
</div>
<div id="working-with-eupathdb-granges-resources" class="section level1">
<h1>Working with EuPathDB GRanges resources</h1>
<p>In addition to retrieving gene annotations, AnnotationHub can also be used to query GenomicRange (GRange) objects containing information about gene and transcript structure.</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb67-1" title="1"><span class="co"># query AnnotationHub</span></a>
<a class="sourceLine" id="cb67-2" title="2">res &lt;-<span class="st"> </span><span class="kw">query</span>(ah, <span class="kw">c</span>(<span class="st">'Leishmania major strain Friedlin'</span>, <span class="st">'GRanges'</span>, <span class="st">'EuPathDB'</span>))</a>
<a class="sourceLine" id="cb67-3" title="3">res</a></code></pre></div>
<pre><code>## AnnotationHub with 1 record
## # snapshotDate(): 2018-10-24 
## # names(): AH65354
## # $dataprovider: TriTrypDB
## # $species: Leishmania major
## # $rdataclass: GRanges
## # $rdatadateadded: 2018-10-15
## # $title: Leishmania major strain Friedlin transcript information
## # $description: TriTrypDB 39 transcript information for Leishmania majo...
## # $taxonomyid: 347515
## # $genome: TriTrypDB-39_LmajorFriedlin
## # $sourcetype: GFF
## # $sourceurl: http://TriTrypDB.org/common/downloads/Current_Release/Lma...
## # $sourcesize: NA
## # $tags:
## #   c(&quot;Annotation,EuPathDB,Eukaryote,Pathogen,Parasite,Trypanosome,Kinetoplastid,Leishmania&quot;,
## #   &quot;EuPathDB&quot;) 
## # retrieve record with 'object[[&quot;AH65354&quot;]]'</code></pre>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb69-1" title="1"><span class="co"># retrieve a GRanges instance associated with the result record</span></a>
<a class="sourceLine" id="cb69-2" title="2">gr &lt;-<span class="st"> </span>res[[<span class="st">'AH65354'</span>]]</a></code></pre></div>
<pre><code>## downloading 0 resources</code></pre>
<pre><code>## loading from cache 
##     '/home/trey//.AnnotationHub/72100'</code></pre>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb72-1" title="1">gr</a></code></pre></div>
<pre><code>## GRanges object with 37311 ranges and 9 metadata columns:
##           seqnames        ranges strand |   source     type     score
##              &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;
##       [1]  LmjF.24 117748-119043      + | EuPathDB     gene      &lt;NA&gt;
##       [2]  LmjF.24 117748-119043      + | EuPathDB     mRNA      &lt;NA&gt;
##       [3]  LmjF.24 117748-119043      + | EuPathDB     exon      &lt;NA&gt;
##       [4]  LmjF.24 117748-119043      + | EuPathDB      CDS      &lt;NA&gt;
##       [5]  LmjF.34 961963-966600      - | EuPathDB     gene      &lt;NA&gt;
##       ...      ...           ...    ... .      ...      ...       ...
##   [37307]  LmjF.11 425050-425856      + | EuPathDB      CDS      &lt;NA&gt;
##   [37308]  LmjF.27 718772-719071      - | EuPathDB     gene      &lt;NA&gt;
##   [37309]  LmjF.27 718772-719071      - | EuPathDB     mRNA      &lt;NA&gt;
##   [37310]  LmjF.27 718772-719071      - | EuPathDB     exon      &lt;NA&gt;
##   [37311]  LmjF.27 718772-719071      - | EuPathDB      CDS      &lt;NA&gt;
##               phase                        ID
##           &lt;integer&gt;               &lt;character&gt;
##       [1]      &lt;NA&gt;              LmjF.24.0370
##       [2]      &lt;NA&gt;         LmjF.24.0370:mRNA
##       [3]      &lt;NA&gt;      exon_LmjF.24.0370-E1
##       [4]         0 LmjF.24.0370:mRNA-p1-CDS1
##       [5]      &lt;NA&gt;              LmjF.34.2180
##       ...       ...                       ...
##   [37307]         0 LmjF.11.1030:mRNA-p1-CDS1
##   [37308]      &lt;NA&gt;              LmjF.27.1740
##   [37309]      &lt;NA&gt;         LmjF.27.1740:mRNA
##   [37310]      &lt;NA&gt;      exon_LmjF.27.1740-E1
##   [37311]         0 LmjF.27.1740:mRNA-p1-CDS1
##                                      description            Parent
##                                      &lt;character&gt;   &lt;CharacterList&gt;
##       [1]   aspartate aminotransferase, putative              &lt;NA&gt;
##       [2]   aspartate aminotransferase, putative      LmjF.24.0370
##       [3]                                   &lt;NA&gt; LmjF.24.0370:mRNA
##       [4]                                   &lt;NA&gt; LmjF.24.0370:mRNA
##       [5]        hypothetical protein, conserved              &lt;NA&gt;
##       ...                                    ...               ...
##   [37307]                                   &lt;NA&gt; LmjF.11.1030:mRNA
##   [37308] hypothetical protein, unknown function              &lt;NA&gt;
##   [37309] hypothetical protein, unknown function      LmjF.27.1740
##   [37310]                                   &lt;NA&gt; LmjF.27.1740:mRNA
##   [37311]                                   &lt;NA&gt; LmjF.27.1740:mRNA
##                                       Note    protein_source_id
##                            &lt;CharacterList&gt;          &lt;character&gt;
##       [1]                             &lt;NA&gt;                 &lt;NA&gt;
##       [2] 2.6.1.1 (Aspartate transaminase)                 &lt;NA&gt;
##       [3]                             &lt;NA&gt;                 &lt;NA&gt;
##       [4]                             &lt;NA&gt; LmjF.24.0370:mRNA-p1
##       [5]                             &lt;NA&gt;                 &lt;NA&gt;
##       ...                              ...                  ...
##   [37307]                             &lt;NA&gt; LmjF.11.1030:mRNA-p1
##   [37308]                             &lt;NA&gt;                 &lt;NA&gt;
##   [37309]                             &lt;NA&gt;                 &lt;NA&gt;
##   [37310]                             &lt;NA&gt;                 &lt;NA&gt;
##   [37311]                             &lt;NA&gt; LmjF.27.1740:mRNA-p1
##   -------
##   seqinfo: 36 sequences from TriTrypDB-39_LmajorFriedlin genome; no seqlengths</code></pre>
<p>The resulting <code>GRanges</code> object can then be interacted with using the <a href="https://bioconductor.org/packages/3.7/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.pdf">standard GRanges functions</a>, including:</p>
<ul>
<li>seqnames</li>
<li>strand</li>
<li>width</li>
</ul>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb74-1" title="1"><span class="co"># chromosome names</span></a>
<a class="sourceLine" id="cb74-2" title="2"><span class="kw">seqnames</span>(gr)</a></code></pre></div>
<pre><code>## factor-Rle of length 37311 with 9006 runs
##   Lengths:       4       4       4       4 ...       4       4       4
##   Values : LmjF.24 LmjF.34 LmjF.36 LmjF.28 ... LmjF.04 LmjF.11 LmjF.27
## Levels(36): LmjF.01 LmjF.02 LmjF.03 LmjF.04 ... LmjF.34 LmjF.35 LmjF.36</code></pre>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb76-1" title="1"><span class="co"># strand information</span></a>
<a class="sourceLine" id="cb76-2" title="2"><span class="kw">strand</span>(gr)</a></code></pre></div>
<pre><code>## factor-Rle of length 37311 with 4630 runs
##   Lengths:  4 12  4  4  4  4  3  4 31  4 ...  8  8  4  4  4  4  4 23  4  4
##   Values :  +  -  +  -  +  -  +  -  +  - ...  +  -  +  -  +  -  +  -  +  -
## Levels(3): + - *</code></pre>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb78-1" title="1"><span class="co"># feature widths</span></a>
<a class="sourceLine" id="cb78-2" title="2"><span class="kw">head</span>(<span class="kw">width</span>(gr))</a></code></pre></div>
<pre><code>## [1] 1296 1296 1296 1296 4638 4638</code></pre>
<p>Some information can be retrieve directly as object properties using the <code>$</code> operator:</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb80-1" title="1"><span class="co"># list of location types in the resource</span></a>
<a class="sourceLine" id="cb80-2" title="2"><span class="kw">table</span>(gr<span class="op">$</span>type)</a></code></pre></div>
<pre><code>## 
##            gene            mRNA            exon             CDS 
##            9378            8519            9583            8606 
##          snoRNA            rRNA three_prime_UTR  five_prime_UTR 
##             741              63               8             241 
##           ncRNA            tRNA           snRNA 
##              83              83               6</code></pre>
<p>To subset the GRanges instance, you can use the standard <code>[</code> operator:</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb82-1" title="1"><span class="co"># get the first three ranges</span></a>
<a class="sourceLine" id="cb82-2" title="2">gr[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 9 metadata columns:
##       seqnames        ranges strand |   source     type     score
##          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;
##   [1]  LmjF.24 117748-119043      + | EuPathDB     gene      &lt;NA&gt;
##   [2]  LmjF.24 117748-119043      + | EuPathDB     mRNA      &lt;NA&gt;
##   [3]  LmjF.24 117748-119043      + | EuPathDB     exon      &lt;NA&gt;
##           phase                   ID                          description
##       &lt;integer&gt;          &lt;character&gt;                          &lt;character&gt;
##   [1]      &lt;NA&gt;         LmjF.24.0370 aspartate aminotransferase, putative
##   [2]      &lt;NA&gt;    LmjF.24.0370:mRNA aspartate aminotransferase, putative
##   [3]      &lt;NA&gt; exon_LmjF.24.0370-E1                                 &lt;NA&gt;
##                  Parent                             Note protein_source_id
##         &lt;CharacterList&gt;                  &lt;CharacterList&gt;       &lt;character&gt;
##   [1]              &lt;NA&gt;                             &lt;NA&gt;              &lt;NA&gt;
##   [2]      LmjF.24.0370 2.6.1.1 (Aspartate transaminase)              &lt;NA&gt;
##   [3] LmjF.24.0370:mRNA                             &lt;NA&gt;              &lt;NA&gt;
##   -------
##   seqinfo: 36 sequences from TriTrypDB-39_LmajorFriedlin genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb84-1" title="1"><span class="co"># get all gene entries on chromosome 4</span></a>
<a class="sourceLine" id="cb84-2" title="2">gr[gr<span class="op">$</span>type <span class="op">==</span><span class="st"> 'gene'</span> <span class="op">&amp;</span><span class="st"> </span><span class="kw">seqnames</span>(gr) <span class="op">==</span><span class="st"> 'LmjF.04'</span>]</a></code></pre></div>
<pre><code>## GRanges object with 130 ranges and 9 metadata columns:
##         seqnames        ranges strand |   source     type     score
##            &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;
##     [1]  LmjF.04 418659-420857      - | EuPathDB     gene      &lt;NA&gt;
##     [2]  LmjF.04 185923-186489      - | EuPathDB     gene      &lt;NA&gt;
##     [3]  LmjF.04 283441-286785      - | EuPathDB     gene      &lt;NA&gt;
##     [4]  LmjF.04 421926-422858      - | EuPathDB     gene      &lt;NA&gt;
##     [5]  LmjF.04 124169-125230      + | EuPathDB     gene      &lt;NA&gt;
##     ...      ...           ...    ... .      ...      ...       ...
##   [126]  LmjF.04 181903-185242      - | EuPathDB     gene      &lt;NA&gt;
##   [127]  LmjF.04 375010-376536      - | EuPathDB     gene      &lt;NA&gt;
##   [128]  LmjF.04 275783-276241      - | EuPathDB     gene      &lt;NA&gt;
##   [129]  LmjF.04 254218-254724      - | EuPathDB     gene      &lt;NA&gt;
##   [130]  LmjF.04 440646-441581      - | EuPathDB     gene      &lt;NA&gt;
##             phase           ID
##         &lt;integer&gt;  &lt;character&gt;
##     [1]      &lt;NA&gt; LmjF.04.1120
##     [2]      &lt;NA&gt; LmjF.04.0470
##     [3]      &lt;NA&gt; LmjF.04.0700
##     [4]      &lt;NA&gt; LmjF.04.1130
##     [5]      &lt;NA&gt; LmjF.04.0380
##     ...       ...          ...
##   [126]      &lt;NA&gt; LmjF.04.0465
##   [127]      &lt;NA&gt; LmjF.04.0970
##   [128]      &lt;NA&gt; LmjF.04.0670
##   [129]      &lt;NA&gt; LmjF.04.0610
##   [130]      &lt;NA&gt; LmjF.04.1170
##                                            description          Parent
##                                            &lt;character&gt; &lt;CharacterList&gt;
##     [1]                hypothetical protein, conserved            &lt;NA&gt;
##     [2]            60S ribosomal protein L11 (L5, L16)            &lt;NA&gt;
##     [3]                hypothetical protein, conserved            &lt;NA&gt;
##     [4] cytochrome c oxidase assembly factor, putative            &lt;NA&gt;
##     [5]     phytanoyl-CoA dioxygenase (PhyH), putative            &lt;NA&gt;
##     ...                                            ...             ...
##   [126]                           hypothetical protein            &lt;NA&gt;
##   [127]                hypothetical protein, conserved            &lt;NA&gt;
##   [128]                hypothetical protein, conserved            &lt;NA&gt;
##   [129]                hypothetical protein, conserved            &lt;NA&gt;
##   [130]            Double RNA binding domain protein 3            &lt;NA&gt;
##                    Note protein_source_id
##         &lt;CharacterList&gt;       &lt;character&gt;
##     [1]            &lt;NA&gt;              &lt;NA&gt;
##     [2]            &lt;NA&gt;              &lt;NA&gt;
##     [3]            &lt;NA&gt;              &lt;NA&gt;
##     [4]            &lt;NA&gt;              &lt;NA&gt;
##     [5]            &lt;NA&gt;              &lt;NA&gt;
##     ...             ...               ...
##   [126]            &lt;NA&gt;              &lt;NA&gt;
##   [127]            &lt;NA&gt;              &lt;NA&gt;
##   [128]            &lt;NA&gt;              &lt;NA&gt;
##   [129]            &lt;NA&gt;              &lt;NA&gt;
##   [130]            &lt;NA&gt;              &lt;NA&gt;
##   -------
##   seqinfo: 36 sequences from TriTrypDB-39_LmajorFriedlin genome; no seqlengths</code></pre>
</div>
<div id="installing-a-new-species-in-your-r-session" class="section level1">
<h1>Installing a new species in your R session</h1>
<p>When AnnotationHub and the EupathDB web resources get out of sync, one might want to install and manipulate the newest version of the available Eupath data. The following blocks demonstrate how one might do those tasks.</p>
<div id="getting-a-less-popular-tritrypdb-data-set" class="section level2">
<h2>Getting a less popular TriTrypDB data set</h2>
<p>In some of my work, it is interesting to compare genes from species like Leishmania to Crithidia fasculata. In order to get the newest version of that data, we need to query the TriTrypDB web resource rather than EupathDB.</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb86-1" title="1"><span class="co">## Note that some but not all web services have moved to https...</span></a>
<a class="sourceLine" id="cb86-2" title="2">tri_meta &lt;-<span class="st"> </span><span class="kw">download_eupath_metadata</span>(<span class="dt">webservice=</span><span class="st">&quot;tritrypdb&quot;</span>)</a></code></pre></div>
<pre><code>## Warning in download.file(url = request_url, destfile = metadata_json):
## cannot open URL 'https://tritrypdb.org/tritrypdb/webservices/
## OrganismQuestions/GenomeDataTypes.json?o-fields=all': HTTP status was '404
## Not Found'</code></pre>
<pre><code>## Downloaded: http://tritrypdb.org/tritrypdb/webservices/OrganismQuestions/GenomeDataTypes.json?o-fields=all</code></pre>
<pre><code>## Warning: 'BiocInstaller' and 'biocLite()' are deprecated, use the 'BiocManager'
##   CRAN package instead.</code></pre>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb90-1" title="1">crit_name &lt;-<span class="st"> </span><span class="kw">check_eupath_species</span>(<span class="dt">species=</span><span class="st">&quot;Crith&quot;</span>, <span class="dt">metadata=</span>tri_meta)</a>
<a class="sourceLine" id="cb90-2" title="2">crit_name</a></code></pre></div>
<pre><code>##    BiocVersion                        Genome NumGenes NumOrthologs
## 24         3.8 TriTrypDB-41_CfasciculataCfCl     9619         9150
##    SourceType
## 24        GFF
##                                                                                                       SourceUrl
## 24 http://TriTrypDB.org/common/downloads/release-41/CfasciculataCfCl/gff/data/TriTrypDB-41_CfasciculataCfCl.gff
##    SourceVersion                            Species TaxonomyId
## 24            41 Crithidia fasciculata strain Cf-Cl         NA
##    Coordinate_1_based DataProvider                       Maintainer
## 24               TRUE    TriTrypDB Keith Hughitt &lt;khughitt@umd.edu&gt;
##                                                                                    Tags
## 24 Annotation,EuPathDB,Eukaryote,Pathogen,Parasite,Trypanosome,Kinetoplastid,Leishmania
##                                                             Title
## 24 Genome wide annotations for Crithidia fasciculata strain Cf-Cl
##                                                        Description
## 24 TriTrypDB 41 annotations for Crithidia fasciculata strain Cf-Cl
##    RDataClass DispatchClass
## 24      OrgDb    SQLiteFile
##                                                ResourceName
## 24 org.Crithidia_fasciculata_strain_Cf-Cl.tritryp.db.sqlite
##                                                                      RDataPath
## 24 EuPathDB/OrgDb/3.8/org.Crithidia_fasciculata_strain_Cf-Cl.tritryp.db.sqlite</code></pre>
<div id="creating-the-package" class="section level3">
<h3>Creating the package</h3>
<p>Now that we have the canonical name from the TriTrypDB for Crithidia, we can create a fresh orgdb package for it. I will not actually run this because it takes a long time and prints a lot to screen.</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb92-1" title="1">orgdb_pkg &lt;-<span class="st"> </span><span class="kw">make_eupath_orgdb</span>(<span class="dt">species=</span>crit_name[[<span class="st">&quot;Species&quot;</span>]], <span class="dt">entry=</span>crit_name, <span class="dt">reinstall=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb92-2" title="2">txdb_pkg &lt;-<span class="st"> </span><span class="kw">make_eupath_txdb</span>(<span class="dt">species=</span>crit_name[[<span class="st">&quot;Species&quot;</span>]], <span class="dt">entry=</span>crit_name)</a>
<a class="sourceLine" id="cb92-3" title="3">bsgenome_pkg &lt;-<span class="st"> </span><span class="kw">make_eupath_bsgenome</span>(<span class="dt">species=</span>crit_name[[<span class="st">&quot;Species&quot;</span>]], <span class="dt">entry=</span>crit_name)</a>
<a class="sourceLine" id="cb92-4" title="4">organ_pkg &lt;-<span class="st"> </span><span class="kw">make_eupath_organismdbi</span>(<span class="dt">species=</span>crit_name[[<span class="st">&quot;Species&quot;</span>]], <span class="dt">entry=</span>crit_name)</a></code></pre></div>
</div>
</div>
<div id="extracting-annotation-data" class="section level2">
<h2>Extracting annotation data</h2>
<p>The following demonstrates ways to extract data from the generated orgdb package.</p>
<div id="choosing-and-loading-a-generated-package" class="section level3">
<h3>Choosing and loading a generated package</h3>
<p>Because the eupathdb is constantly evolving, the get_eupath_pkgnames() function will use the metadata generated in download_eupath_metadata() in order to provide a differently names package for each eupathdb revision.</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb93-1" title="1">orgdb_pkg &lt;-<span class="st"> </span><span class="kw">get_eupath_pkgnames</span>(crit_name[[<span class="st">&quot;Species&quot;</span>]])[[<span class="st">&quot;orgdb&quot;</span>]]</a></code></pre></div>
<pre><code>## Starting metadata download.</code></pre>
<pre><code>## Downloaded: https://eupathdb.org/eupathdb/webservices/OrganismQuestions/GenomeDataTypes.json?o-fields=all</code></pre>
<pre><code>## Finished metadata download.</code></pre>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb97-1" title="1"><span class="co">## Get the name of the Crithidia package that got created by make_eupath_orgdb.</span></a>
<a class="sourceLine" id="cb97-2" title="2">orgdb_pkg</a></code></pre></div>
<pre><code>## org.Cfasciculata.Cf.Cl.v41.eg.db</code></pre>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb99-1" title="1"><span class="co">## Thus we see the v41 (as of late 2018), a number which presumably will continue increasing.</span></a>
<a class="sourceLine" id="cb99-2" title="2"><span class="co">## We can set the version parameter to change this if we have a previous version installed.</span></a>
<a class="sourceLine" id="cb99-3" title="3"></a>
<a class="sourceLine" id="cb99-4" title="4"><span class="co">## Now get the set of available columns from it:</span></a>
<a class="sourceLine" id="cb99-5" title="5"><span class="kw">library</span>(orgdb_pkg, <span class="dt">character=</span><span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## </code></pre>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb101-1" title="1">avail_columns &lt;-<span class="st"> </span>AnnotationDbi<span class="op">::</span><span class="kw">columns</span>(<span class="kw">get0</span>(orgdb_pkg))</a>
<a class="sourceLine" id="cb101-2" title="2">avail_columns</a></code></pre></div>
<pre><code>##   [1] &quot;ANNOT_ANNOTATED_GO_COMPONENT&quot;           
##   [2] &quot;ANNOT_ANNOTATED_GO_FUNCTION&quot;            
##   [3] &quot;ANNOT_ANNOTATED_GO_ID_COMPONENT&quot;        
##   [4] &quot;ANNOT_ANNOTATED_GO_ID_FUNCTION&quot;         
##   [5] &quot;ANNOT_ANNOTATED_GO_ID_PROCESS&quot;          
##   [6] &quot;ANNOT_ANNOTATED_GO_PROCESS&quot;             
##   [7] &quot;ANNOT_CDS&quot;                              
##   [8] &quot;ANNOT_CDS_LENGTH&quot;                       
##   [9] &quot;ANNOT_CHROMOSOME&quot;                       
##  [10] &quot;ANNOT_EC_NUMBERS&quot;                       
##  [11] &quot;ANNOT_EC_NUMBERS_DERIVED&quot;               
##  [12] &quot;ANNOT_EXON_COUNT&quot;                       
##  [13] &quot;ANNOT_FIVE_PRIME_UTR_LENGTH&quot;            
##  [14] &quot;ANNOT_GENE_ENTREZ_ID&quot;                   
##  [15] &quot;ANNOT_GENE_EXON_COUNT&quot;                  
##  [16] &quot;ANNOT_GENE_HTS_NONCODING_SNPS&quot;          
##  [17] &quot;ANNOT_GENE_HTS_NONSYN_SYN_RATIO&quot;        
##  [18] &quot;ANNOT_GENE_HTS_NONSYNONYMOUS_SNPS&quot;      
##  [19] &quot;ANNOT_GENE_HTS_STOP_CODON_SNPS&quot;         
##  [20] &quot;ANNOT_GENE_HTS_SYNONYMOUS_SNPS&quot;         
##  [21] &quot;ANNOT_GENE_LOCATION_TEXT&quot;               
##  [22] &quot;ANNOT_GENE_NAME&quot;                        
##  [23] &quot;ANNOT_GENE_ORTHOLOG_NUMBER&quot;             
##  [24] &quot;ANNOT_GENE_ORTHOMCL_NAME&quot;               
##  [25] &quot;ANNOT_GENE_PARALOG_NUMBER&quot;              
##  [26] &quot;ANNOT_GENE_PREVIOUS_IDS&quot;                
##  [27] &quot;ANNOT_GENE_PRODUCT&quot;                     
##  [28] &quot;ANNOT_GENE_SOURCE_ID&quot;                   
##  [29] &quot;ANNOT_GENE_TOTAL_HTS_SNPS&quot;              
##  [30] &quot;ANNOT_GENE_TRANSCRIPT_COUNT&quot;            
##  [31] &quot;ANNOT_GENE_TYPE&quot;                        
##  [32] &quot;ANNOT_HAS_MISSING_TRANSCRIPTS&quot;          
##  [33] &quot;ANNOT_INTERPRO_DESCRIPTION&quot;             
##  [34] &quot;ANNOT_INTERPRO_ID&quot;                      
##  [35] &quot;ANNOT_IS_PSEUDO&quot;                        
##  [36] &quot;ANNOT_ISOELECTRIC_POINT&quot;                
##  [37] &quot;ANNOT_LOCATION_TEXT&quot;                    
##  [38] &quot;ANNOT_MATCHED_RESULT&quot;                   
##  [39] &quot;ANNOT_MOLECULAR_WEIGHT&quot;                 
##  [40] &quot;ANNOT_ORGANISM&quot;                         
##  [41] &quot;ANNOT_PFAM_DESCRIPTION&quot;                 
##  [42] &quot;ANNOT_PFAM_ID&quot;                          
##  [43] &quot;ANNOT_PIRSF_DESCRIPTION&quot;                
##  [44] &quot;ANNOT_PIRSF_ID&quot;                         
##  [45] &quot;ANNOT_PREDICTED_GO_COMPONENT&quot;           
##  [46] &quot;ANNOT_PREDICTED_GO_FUNCTION&quot;            
##  [47] &quot;ANNOT_PREDICTED_GO_ID_COMPONENT&quot;        
##  [48] &quot;ANNOT_PREDICTED_GO_ID_FUNCTION&quot;         
##  [49] &quot;ANNOT_PREDICTED_GO_ID_PROCESS&quot;          
##  [50] &quot;ANNOT_PREDICTED_GO_PROCESS&quot;             
##  [51] &quot;ANNOT_PROJECT_ID&quot;                       
##  [52] &quot;ANNOT_PROSITEPROFILES_DESCRIPTION&quot;      
##  [53] &quot;ANNOT_PROSITEPROFILES_ID&quot;               
##  [54] &quot;ANNOT_PROTEIN_LENGTH&quot;                   
##  [55] &quot;ANNOT_PROTEIN_SEQUENCE&quot;                 
##  [56] &quot;ANNOT_SEQUENCE_ID&quot;                      
##  [57] &quot;ANNOT_SIGNALP_PEPTIDE&quot;                  
##  [58] &quot;ANNOT_SIGNALP_SCORES&quot;                   
##  [59] &quot;ANNOT_SMART_DESCRIPTION&quot;                
##  [60] &quot;ANNOT_SMART_ID&quot;                         
##  [61] &quot;ANNOT_SOURCE_ID&quot;                        
##  [62] &quot;ANNOT_STRAND&quot;                           
##  [63] &quot;ANNOT_SUPERFAMILY_DESCRIPTION&quot;          
##  [64] &quot;ANNOT_SUPERFAMILY_ID&quot;                   
##  [65] &quot;ANNOT_THREE_PRIME_UTR_LENGTH&quot;           
##  [66] &quot;ANNOT_TIGRFAM_DESCRIPTION&quot;              
##  [67] &quot;ANNOT_TIGRFAM_ID&quot;                       
##  [68] &quot;ANNOT_TM_COUNT&quot;                         
##  [69] &quot;ANNOT_TRANS_FOUND_PER_GENE_INTERNAL&quot;    
##  [70] &quot;ANNOT_TRANSCRIPT_INDEX_PER_GENE&quot;        
##  [71] &quot;ANNOT_TRANSCRIPT_LENGTH&quot;                
##  [72] &quot;ANNOT_TRANSCRIPT_LINK&quot;                  
##  [73] &quot;ANNOT_TRANSCRIPT_PRODUCT&quot;               
##  [74] &quot;ANNOT_TRANSCRIPT_SEQUENCE&quot;              
##  [75] &quot;ANNOT_TRANSCRIPTS_FOUND_PER_GENE&quot;       
##  [76] &quot;ANNOT_UNIPROT_ID&quot;                       
##  [77] &quot;ANNOT_URI&quot;                              
##  [78] &quot;ANNOT_WDK_WEIGHT&quot;                       
##  [79] &quot;CHR_ID&quot;                                 
##  [80] &quot;GENE_TYPE&quot;                              
##  [81] &quot;GID&quot;                                    
##  [82] &quot;GO_EVIDENCE_CODE&quot;                       
##  [83] &quot;GO_GO_ID&quot;                               
##  [84] &quot;GO_GO_TERM_NAME&quot;                        
##  [85] &quot;GO_IS_NOT&quot;                              
##  [86] &quot;GO_ONTOLOGY&quot;                            
##  [87] &quot;GO_REFERENCE&quot;                           
##  [88] &quot;GO_SORT_KEY&quot;                            
##  [89] &quot;GO_SOURCE&quot;                              
##  [90] &quot;GO_SUPPORT_FOR_EVIDENCE_CODE_ASSIGNMENT&quot;
##  [91] &quot;GO_TRANSCRIPT_ID_S_&quot;                    
##  [92] &quot;INTERPRO_DESCRIPTION&quot;                   
##  [93] &quot;INTERPRO_E_VALUE&quot;                       
##  [94] &quot;INTERPRO_END_MIN&quot;                       
##  [95] &quot;INTERPRO_INTERPRO_ID&quot;                   
##  [96] &quot;INTERPRO_NAME&quot;                          
##  [97] &quot;INTERPRO_PRIMARY_ID&quot;                    
##  [98] &quot;INTERPRO_SECONDARY_ID&quot;                  
##  [99] &quot;INTERPRO_START_MIN&quot;                     
## [100] &quot;INTERPRO_TRANSCRIPT_ID_S_&quot;              
## [101] &quot;ORGANISM&quot;                               
## [102] &quot;ORTHOLOG_COUNT&quot;                         
## [103] &quot;ORTHOLOG_GROUP&quot;                         
## [104] &quot;ORTHOLOG_ID&quot;                            
## [105] &quot;PARALOG_COUNT&quot;                          
## [106] &quot;PATHWAY_EC_NUMBER_MATCHED_IN_PATHWAY&quot;   
## [107] &quot;PATHWAY_EXACT_EC_NUMBER_MATCH&quot;          
## [108] &quot;PATHWAY_EXPASY_URL&quot;                     
## [109] &quot;PATHWAY_PATHWAY&quot;                        
## [110] &quot;PATHWAY_PATHWAY_SOURCE&quot;                 
## [111] &quot;PATHWAY_PATHWAY_SOURCE_ID&quot;              
## [112] &quot;PATHWAY_REACTIONS_MATCHING_EC_NUMBER&quot;</code></pre>
</div>
<div id="extracting-data-from-a-package" class="section level3">
<h3>Extracting data from a package</h3>
<p>Once we have a package loaded, everything else is an application of the AnnotationDbi interface. Here are a few examples.</p>
<p>Later I will show a couple of shortcuts for these methods.</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb103-1" title="1"><span class="co">## The columns which begin with strings like 'PATHWAY' or 'INTERPRO' are actually separate</span></a>
<a class="sourceLine" id="cb103-2" title="2"><span class="co">## sql tables in the orgdb database, and as such will lead to a hugely redundant data table</span></a>
<a class="sourceLine" id="cb103-3" title="3"><span class="co">## if we select them.</span></a>
<a class="sourceLine" id="cb103-4" title="4"></a>
<a class="sourceLine" id="cb103-5" title="5">chosen_columns_idx &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="dt">x=</span>avail_columns, <span class="dt">pattern=</span><span class="st">&quot;^ANNOT&quot;</span>)</a>
<a class="sourceLine" id="cb103-6" title="6">chosen_columns &lt;-<span class="st"> </span>avail_columns[chosen_columns_idx]</a>
<a class="sourceLine" id="cb103-7" title="7"></a>
<a class="sourceLine" id="cb103-8" title="8"><span class="co">## Now we have a set of columns of interest, let us get a data table/data frame.</span></a>
<a class="sourceLine" id="cb103-9" title="9">crit_annot &lt;-<span class="st"> </span><span class="kw">load_orgdb_annotations</span>(<span class="dt">orgdb=</span>orgdb_pkg, <span class="dt">keytype=</span><span class="st">&quot;gid&quot;</span>, <span class="dt">fields=</span>chosen_columns)</a></code></pre></div>
<pre><code>## Unable to find CDSNAME, setting it to ANNOT_GENE_NAME.</code></pre>
<pre><code>## Unable to find CDSCHROM in the db, removing it.</code></pre>
<pre><code>## Unable to find CDSSTRAND in the db, removing it.</code></pre>
<pre><code>## Unable to find CDSSTART in the db, removing it.</code></pre>
<pre><code>## Unable to find CDSEND in the db, removing it.</code></pre>
<pre><code>## Extracted all gene ids.</code></pre>
<pre><code>## Attempting to select: ANNOT_GENE_NAME, GENE_TYPE, ANNOT_ANNOTATED_GO_COMPONENT, ANNOT_ANNOTATED_GO_FUNCTION, ANNOT_ANNOTATED_GO_ID_COMPONENT, ANNOT_ANNOTATED_GO_ID_FUNCTION, ANNOT_ANNOTATED_GO_ID_PROCESS, ANNOT_ANNOTATED_GO_PROCESS, ANNOT_CDS, ANNOT_CDS_LENGTH, ANNOT_CHROMOSOME, ANNOT_EC_NUMBERS, ANNOT_EC_NUMBERS_DERIVED, ANNOT_EXON_COUNT, ANNOT_FIVE_PRIME_UTR_LENGTH, ANNOT_GENE_ENTREZ_ID, ANNOT_GENE_EXON_COUNT, ANNOT_GENE_HTS_NONCODING_SNPS, ANNOT_GENE_HTS_NONSYN_SYN_RATIO, ANNOT_GENE_HTS_NONSYNONYMOUS_SNPS, ANNOT_GENE_HTS_STOP_CODON_SNPS, ANNOT_GENE_HTS_SYNONYMOUS_SNPS, ANNOT_GENE_LOCATION_TEXT, ANNOT_GENE_NAME, ANNOT_GENE_ORTHOLOG_NUMBER, ANNOT_GENE_ORTHOMCL_NAME, ANNOT_GENE_PARALOG_NUMBER, ANNOT_GENE_PREVIOUS_IDS, ANNOT_GENE_PRODUCT, ANNOT_GENE_SOURCE_ID, ANNOT_GENE_TOTAL_HTS_SNPS, ANNOT_GENE_TRANSCRIPT_COUNT, ANNOT_GENE_TYPE, ANNOT_HAS_MISSING_TRANSCRIPTS, ANNOT_INTERPRO_DESCRIPTION, ANNOT_INTERPRO_ID, ANNOT_IS_PSEUDO, ANNOT_ISOELECTRIC_POINT, ANNOT_LOCATION_TEXT, ANNOT_MATCHED_RESULT, ANNOT_MOLECULAR_WEIGHT, ANNOT_ORGANISM, ANNOT_PFAM_DESCRIPTION, ANNOT_PFAM_ID, ANNOT_PIRSF_DESCRIPTION, ANNOT_PIRSF_ID, ANNOT_PREDICTED_GO_COMPONENT, ANNOT_PREDICTED_GO_FUNCTION, ANNOT_PREDICTED_GO_ID_COMPONENT, ANNOT_PREDICTED_GO_ID_FUNCTION, ANNOT_PREDICTED_GO_ID_PROCESS, ANNOT_PREDICTED_GO_PROCESS, ANNOT_PROJECT_ID, ANNOT_PROSITEPROFILES_DESCRIPTION, ANNOT_PROSITEPROFILES_ID, ANNOT_PROTEIN_LENGTH, ANNOT_PROTEIN_SEQUENCE, ANNOT_SEQUENCE_ID, ANNOT_SIGNALP_PEPTIDE, ANNOT_SIGNALP_SCORES, ANNOT_SMART_DESCRIPTION, ANNOT_SMART_ID, ANNOT_SOURCE_ID, ANNOT_STRAND, ANNOT_SUPERFAMILY_DESCRIPTION, ANNOT_SUPERFAMILY_ID, ANNOT_THREE_PRIME_UTR_LENGTH, ANNOT_TIGRFAM_DESCRIPTION, ANNOT_TIGRFAM_ID, ANNOT_TM_COUNT, ANNOT_TRANS_FOUND_PER_GENE_INTERNAL, ANNOT_TRANSCRIPT_INDEX_PER_GENE, ANNOT_TRANSCRIPT_LENGTH, ANNOT_TRANSCRIPT_LINK, ANNOT_TRANSCRIPT_PRODUCT, ANNOT_TRANSCRIPT_SEQUENCE, ANNOT_TRANSCRIPTS_FOUND_PER_GENE, ANNOT_UNIPROT_ID, ANNOT_URI, ANNOT_WDK_WEIGHT</code></pre>
<pre><code>## 'select()' returned 1:1 mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb112-1" title="1"><span class="co">## load_orgdb_annotations will fill out separate dataframes for each annotation type,</span></a>
<a class="sourceLine" id="cb112-2" title="2"><span class="co">## genes, exons, transcripts, etc.  In this case, we only want the genes</span></a>
<a class="sourceLine" id="cb112-3" title="3"><span class="co">## (The eupathdb does not provide much information for the others.)</span></a>
<a class="sourceLine" id="cb112-4" title="4">crit_genes &lt;-<span class="st"> </span>crit_annot[[<span class="st">&quot;genes&quot;</span>]]</a>
<a class="sourceLine" id="cb112-5" title="5"><span class="kw">dim</span>(crit_genes)</a></code></pre></div>
<pre><code>## [1] 9489   80</code></pre>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb114-1" title="1"><span class="kw">head</span>(crit_genes)</a></code></pre></div>
<pre><code>##                             gid annot_gene_name      gene_type
## CFAC1_010005600 CFAC1_010005600                 protein coding
## CFAC1_010005700 CFAC1_010005700                 protein coding
## CFAC1_010005800 CFAC1_010005800                 protein coding
## CFAC1_010005900 CFAC1_010005900                 protein coding
## CFAC1_010006000 CFAC1_010006000                 protein coding
## CFAC1_010006100 CFAC1_010006100                 protein coding
##                 annot_annotated_go_component annot_annotated_go_function
## CFAC1_010005600                                                         
## CFAC1_010005700                                                         
## CFAC1_010005800                                                         
## CFAC1_010005900                                                         
## CFAC1_010006000                                                         
## CFAC1_010006100                                                         
##                 annot_annotated_go_id_component
## CFAC1_010005600                                
## CFAC1_010005700                                
## CFAC1_010005800                                
## CFAC1_010005900                                
## CFAC1_010006000                                
## CFAC1_010006100                                
##                 annot_annotated_go_id_function
## CFAC1_010005600                               
## CFAC1_010005700                               
## CFAC1_010005800                               
## CFAC1_010005900                               
## CFAC1_010006000                               
## CFAC1_010006100                               
##                 annot_annotated_go_id_process annot_annotated_go_process
## CFAC1_010005600                                                         
## CFAC1_010005700                                                         
## CFAC1_010005800                                                         
## CFAC1_010005900                                                         
## CFAC1_010006000                                                         
## CFAC1_010006100                                                         
##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    annot_cds
## CFAC1_010005600                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   ATGGACGTGCCAAGCGAAGATTATGACGTGAGTATGATGATAAAAGCCATCAACGAGGTGCAGGCAATTCTCACCTCGCGTCCGTTTGCGACGTTCTCGGAAGTCGATACCGAGATCGCCAAGACGCTGAAACGGTACGAGGCGTTTGGCGACGGCTTTGAGCGCTTTCACGTGAACCTCACAAAAGACGCACTCCAGTCGAATGACTTACAGAAGAGTTTGAAAGACATGAACAAGCGCTGCCAAGAGCGGCTGCGAGACTGTGAATGCAGTCAAAAGGACCAAATTAACGATATCCTTCCTTTTGTTCGCAGTACGTCTTCCATTCTCGTCCACGGCTCTGGCAACCTTTTAGCGCTTGCTATTGCTTGTGCCATTCAGGAGCACGAGGGCGTTCGCTTCTACGTTTGTGAAGGCCGCCCGGCACGCAAAGGCTTCCCGAACGGCTCAGGTGAGAAGCTGCTAGAGAAGGTCGCTGCGACGCCAGAGGGAATGCGTATCAAGGACAGGCTTCACAAGTACTGCACAATTGTTCCAGACAGTGGTGTCAGCTCCGTGATGAATGCAGTTGACTTTGTAGTAATGGGAGCCAACTGTGTCACTGAGCACGGTGGTCTTGTGCACTCCACCGGGTCTCTGCAGATCGCCATCGTTGCCGCGTTTCGCAATGTGCCCTTTTATGTTTTGTGTGAAACCTTCAAGTTTGCGCACATCTTTCCTCTGGGCACAGAGGACCTCAAGCAGCCGGAGGACGGTGCTGGACAGGTTGTCCCGCTTGTGGAGTTTGTCCCTCCATCGATGATCGCTTTGATTTTCTCCGACCAGGGAATTATGCCTCCGTCAGCTGTGGCGGACGAGATGTTTCGCTTCCACACCGCGCCGTCTGCACCATGGAAGCACCCGTAA
## CFAC1_010005700                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     ATGCCGTGCAACTGCCGTCAAGGGACTGCGCTGCACAACCACCTTGGTGACCAGCTCTCCAGCGAAGGCGTGTTTGGCATCGGAGTTTCGATCAACGAAGCAGTCGACCTCGCGGCACTGCAGGTGTGGAACTGCAAGAACAGTGCCCAGGAGGCGGGTCGCCTATTCCATCCGAACCTGGGCGACAACACGTTGCCCATTTGCAGCGATGACGACCAGGAGCTGCTGATTTTCGCACCACTGAGTTCCGTATGTCGTGTCAAGGGCCTGAGCATCAAGGCACCTGCGAGCGAGTGTGCGCCGTCTCGCGTAAAGATCTTTGTGAACATTGTGGACATCACGGGTTTCGGCTCTGTCCAGAGACTGGTGCCACAAGAACAGCTGCAGCTTGCTGACGGCGGCTCAGAAGACCGGATCGTGTACCGCGTGAATCCCGGCAAGTTCTCCTCCGTCTCGTCACTCACGTTCTTCTTCGATGAGAGCTTCAACGGCGACGAGACAGATGTGCTGCGCATCGAGCTCTTTGGAGAGAACACTGGCAAGTCGACGCAGCAGCAGGTCGCCACGAACATTGTGTATGAAGGTCGCGGAAACCCTGCGGACCATCGTACTGCGGAGGAGAAGGAGGCTCTTTTTGAGGTGCGGTGA
## CFAC1_010005800                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            ATGCTCCAGCGTCCCGACCACACTCTCCTGCAGGAGCCGTCGTATCCGAAGGATGTCCAAGCCAAGCTGACGGAGAACGGACCGGCGCAGGCGGGCAAGCAGCTCTTCCAGGCCGACCCTGCCGTCATCGACCCGCAGCTGAGCGAGGCGGTCTCGCTGGGCACCACCATTCTTGCGGTGTCGTACAAGGGCGGTGTCGTGCTCGCCGCTGACTCGCGCACGTCCACTGGTGCGTACGTGGTGAACCGTGTCAGCAACAAGCTCACCAAGCTGACAAAGAAGATCTACTGCTGCCGTAGCGGGTCTGCTGCCGACACGCAGGCCCTAGCGGAGCGCGTCTCCAACTACCTTGGTAGCTACCAGACAGATGTCGGTGCCGACGTCAACGTCGCCACCGCTGCCAACCTGTTCCAGAAGATGTGCTACATGAACAAGTGGAACATCTCCGCTGGTATCATCGTCGCCGGCTACGATCCCATCAACGGCGGCTCTGTCTACAGCATTCCGTCCGGCGGCTCGTGTGTGAAGCTGGACTATGCGCTCGGTGGCAGCGGCTCCATTTTCCTGTACTCGTTCTTCGATGCCAACTATAAGCCTGGCATGACGAAGAGCGAGTGCGTCGCGTTCTGCCAGAGGGCAGTGGCGCACGCGTACAGCCGCGACGGGTCCAGCGGTGGTCTGATCCGGACCATCGCACTCGACGCCGACGAGCCGGAGGACCAGACGGTGCCGTGGAACAAGACGCCGTACTGCATGGAGAAGGACCCCAAGTACCAGGAGCAGGCGGTCCAGAACCCTCCGTTCAGCAGCAGCGCGAAGATTACGGCAAACCGTACGTCGTCGGCGTGA
## CFAC1_010005900 ATGAAGGGGCAGCTGGTCCGGAAAGGCGTAGCGGCGTTGTACACGGCGCCGAAGATTGCGTCGTACACAGATGCTTGCGAGGCACCAACGCCAATGTGGGAGGGTGCGATTCCGCTGAAAGAAGAGCGCATTTTCAAAAACGGTGTTGCGACGCGGATCGTGAAGAAGACGTACTCGCACCCGCCGGAGATTCCACCGGTGCACTTAGGCTTCAACGACGTCAGCTCCATGTACTGTCTCGGCAACGACGAGCTCATGAAGTACTTCCCCGAGGGTCTTGGCGGGAAGGTGATGCAAATGATGCCGCCAGGACACCCGCGCGGCTTCCTGTACCGCGCGCAGAGCCACCTGCTCAACTGCTTCATCGACAAGTTGCCCTTCTGGCAAGCAAAGGAGAAGGCACTCCGCTCCTTAATGAACGGGCGGAGCGGCTTTATCCTTGACGGACCAACCGGCACTGGAAAGAGCGCCATCATGTGTCAAGCAGTCCACTACGCGCGCTCGCGCGACATCCTGACGCTCTACATCCCAAATGCGAAGGCGTGGACGCACGGCGAGTGGTGCTGGCCAAGCACCATCCTTCCTGGCTTTTTCGACGCGCCTGACGCCGCGCGTGAGTTTCTGCGCACGACCGCAGAGGCGAATCGCAGCCTGCTGGAGGGGTGGGACCTGAAGGTGACGCCGCAGGATCTGCCCGTCGAGCAAGGTGAGAAGCTGCCACGGACGCTCTACGACCTGTGCGAGTGGGGCCACCGCGCCGTCGCTCCAGCCTCGATCGATCGTCAGAGCGTGTGCATCAAGTTTTTCCTTGACGAGATTAGCGCCGAAAAGACAAAGCCGATTGTGATCGCGGTGGACGGGTGGAACCTCTTCTCGCACGAGACGCACTTCCGTTACCCGCACCCCGACTTCCTGCGCACGCTCACGACGCTCAACGATGGGTCGACGGACGTCGACCTCTACCCGCAGGAGCTGCCGCGCATTCCGGCGTCGCGTCTCAGCTTCGTGCGCGGGCTCAACAAGCTCATCCTCTCCAAGAACGACCCAAACAAGTTTTTCATTACCTGCACGACGCGCGACTTTAAGCCCTTCGACGGCGTCAGCGGCTTCGCGGATGTGGAGAACGACAGGTTCAAAAACAGCCTCGACGAGTATGCACCGTACGACGCCGAGAAGGACTCCTTCTTCCACCCGATCAAAGTCGACAACTTCAGCGACTACGAGTACCGCTCCTTCCTGCGTTTTGCCATCAACTCGGGCGAACTGGCAGGACTCGGGTGGGGGCCGATGTGGCACTACTCAAGCGACTTTGAGCGGAAGCTGTACAAGATCGGCTTTATGTCGGATCTCAACCCGCAGCGCGTCGTCGATCACTACCACCAGGAGCTTGTGTGGCGCTACGAGTACAAGCGCACCCGCCAGAAGCAGTACCTGCTGGCGCGCCGCAGAATGTGGCAGGATGCGAAGTCGTCGACCGGCGGTCGCGCGGCGAAGCAGTGA
## CFAC1_010006000                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         ATGGCCAAGCGAGAGAAGAAGCGAACGGCTTTGGTGCATCCCAACTTCGACAGTATCAAGCGCTTCTACGCCGTCGCGCCTTTTAGCGCCGCAGAGCGCGCCGCGCTGCCGCCGGTGTACGCCGGTGCCTATAGCCGACGCGCCGCGATGAACGCCTACCACGTCGTTGGCTCGTCACGAGACCAGATCGTGTCAGATGCCATGCGCACTTTCACCCCGGCGTCGTGCACAAAGGAGGTGATGGATCTGTACTTGGAAGTGGCGGCGCTAGTCTCACTGTTTTACGATACGAAGGAGCAGCTCCTTGCGCTGCCGGACATCTAA
## CFAC1_010006100                                                                                                                                                                                                                                                 ATGAGCGATGCGGAGAGAGATCTGGTGGTGTGGAGGGCAGCGGAGCACATCCGGTCCGGCACCTGCGTGGAGCTGCCGCGTGTGGTGCGGACGATCCGCTCTGCAACAGAACACCTGAAGGCGCAGCAAACGAAGTTGGTGGAGAGGGTTTCGGGAAGCACCGTCGTCTTCGGTCCTCTCCGCGGTCACGCCTGCGACTTCGCCAACGTTCTGCTTACCAAGGTCCTGCCCAACCGACAGGTGACAAATCTCGTCTTTCTCGGCAACTACATCGACGGCGCTCACCAGTCTCTCGAGGTGCTCTTCCTCCTTGCCTCTCTCATCGTCTGCTCTCGATTCCGCGTCACTCCGCTTATCGGCAAGCATGAGATGATGTACCCTCTCCAGCCGGAGAACTTCGGCTCGCTGCGCAACGAAATGCTCCTCTTCTGTGCCCGCTCGCCGCACACGTCCCTCGACGACTGCGAGCGCGCCATGCGGGACTTCTTTGCCGTCCTGCCGATCGCCTGTGTGGTGGACAACAAGTACTTCTGTGTAGCTGGAGGACCGGCGTCCGGCTTCCGCACACTAGAGGAGATGGAGGGGAAGCGGGTGACGCAGGACGCGCTGAAGGAGTTCGTGCTGAACGAGCCGATGGACGAGGACGAGGAGCGCATTGCGGACGGCAGTGCTTTTGTGGCGTCGCAGAGCAACGAGCTGGCCTTTCGCTTCACCTTCAACGCGGCGTGCAACTTCTTGAGCCGCAACCGCCTGGCGATGTTTGTGGTAGGCATGGAGTACCACACGAGCCGACCCGATTACGACAGCTTCGCCCGCCCGAACCACTACAAGGAGTCCATCTACTTTCCAGGCTACATGCTGGGCCGCATCCATCCCGACACGCAGGTGCCGGCGGTGCTGACCATGTTCTCCGCACCGTCCTTCTGCGGCGTCAACCGCAACAACGCCTGCATTGCTGAGATCACCGACCGGCGGCTGGAGGTGCAGGAGCTGAGTGTCTACGCTGACCGCCCGCTTGTCACGCCAGGCACGCAGGACCACGCCTTTAGCTGGGCACAGCTGATGCTTGAGCGCGCCGTTGTGGCCATTGCCCGAGAGATTTTGTTTGGTGCCATCAAGGAGGATGCGCCGCCGGAGAGCGATACCCACCACAAGCACCTGGAGGAGATCGCGGTGGCGAAGATGCGGCGCATGTGTCAGCTACTGCAGACGCACGATCTGCCCTTGCCGGAGGTGCCGAAGCTCCCTGTCGACGAGTGA
##                 annot_cds_length annot_chromosome annot_ec_numbers
## CFAC1_010005600            906.0                1                 
## CFAC1_010005700            648.0                1                 
## CFAC1_010005800            849.0                1                 
## CFAC1_010005900           1500.0                1                 
## CFAC1_010006000            324.0                1                 
## CFAC1_010006100           1260.0                1                 
##                                               annot_ec_numbers_derived
## CFAC1_010005600 5.3.1.23 (S-methyl-5-thioribose-1-phosphate isomerase)
## CFAC1_010005700                                                       
## CFAC1_010005800            3.4.25.1 (Proteasome endopeptidase complex)
## CFAC1_010005900                                                       
## CFAC1_010006000                                                       
## CFAC1_010006100                                                       
##                 annot_exon_count annot_five_prime_utr_length
## CFAC1_010005600                1                            
## CFAC1_010005700                1                            
## CFAC1_010005800                1                            
## CFAC1_010005900                1                            
## CFAC1_010006000                1                            
## CFAC1_010006100                1                            
##                 annot_gene_entrez_id annot_gene_exon_count
## CFAC1_010005600                                        1.0
## CFAC1_010005700                                        1.0
## CFAC1_010005800                                        1.0
## CFAC1_010005900                                        1.0
## CFAC1_010006000                                        1.0
## CFAC1_010006100                                        1.0
##                 annot_gene_hts_noncoding_snps
## CFAC1_010005600                           0.0
## CFAC1_010005700                           0.0
## CFAC1_010005800                           0.0
## CFAC1_010005900                           0.0
## CFAC1_010006000                           0.0
## CFAC1_010006100                           0.0
##                 annot_gene_hts_nonsyn_syn_ratio
## CFAC1_010005600                             0.0
## CFAC1_010005700                            0.28
## CFAC1_010005800                             0.0
## CFAC1_010005900                            0.13
## CFAC1_010006000                             0.0
## CFAC1_010006100                            0.17
##                 annot_gene_hts_nonsynonymous_snps
## CFAC1_010005600                               0.0
## CFAC1_010005700                               5.0
## CFAC1_010005800                               0.0
## CFAC1_010005900                               5.0
## CFAC1_010006000                               0.0
## CFAC1_010006100                               3.0
##                 annot_gene_hts_stop_codon_snps
## CFAC1_010005600                            0.0
## CFAC1_010005700                            0.0
## CFAC1_010005800                            0.0
## CFAC1_010005900                            0.0
## CFAC1_010006000                            0.0
## CFAC1_010006100                            0.0
##                 annot_gene_hts_synonymous_snps   annot_gene_location_text
## CFAC1_010005600                           22.0   CfaC1_01:8,824..9,729(+)
## CFAC1_010005700                           18.0 CfaC1_01:10,274..10,921(+)
## CFAC1_010005800                            9.0 CfaC1_01:11,550..12,398(+)
## CFAC1_010005900                           39.0 CfaC1_01:13,166..14,665(+)
## CFAC1_010006000                            9.0 CfaC1_01:15,474..15,797(+)
## CFAC1_010006100                           18.0 CfaC1_01:16,133..17,392(+)
##                 annot_gene_ortholog_number annot_gene_orthomcl_name
## CFAC1_010005600                       39.0               OG5_127805
## CFAC1_010005700                       36.0               OG5_128642
## CFAC1_010005800                       39.0               OG5_127762
## CFAC1_010005900                       40.0               OG5_139168
## CFAC1_010006000                       34.0               OG5_145889
## CFAC1_010006100                       32.0               OG5_154536
##                 annot_gene_paralog_number
## CFAC1_010005600                       0.0
## CFAC1_010005700                       0.0
## CFAC1_010005800                       0.0
## CFAC1_010005900                       0.0
## CFAC1_010006000                       0.0
## CFAC1_010006100                       0.0
##                                                                                      annot_gene_previous_ids
## CFAC1_010005600 Previous IDs: CfaC1_06_0020;CrfA_susu_12.0002;maker-CrfA_susu_12-pred_gff_Automagi-gene-0.50
## CFAC1_010005700 Previous IDs: CfaC1_06_0030;CrfA_susu_12.0003;maker-CrfA_susu_12-pred_gff_Automagi-gene-0.62
## CFAC1_010005800 Previous IDs: CfaC1_06_0040;CrfA_susu_12.0004;maker-CrfA_susu_12-pred_gff_Automagi-gene-0.29
## CFAC1_010005900 Previous IDs: CfaC1_06_0050;CrfA_susu_12.0005;maker-CrfA_susu_12-pred_gff_Automagi-gene-0.51
## CFAC1_010006000 Previous IDs: CfaC1_06_0060;CrfA_susu_12.0006;maker-CrfA_susu_12-pred_gff_Automagi-gene-0.31
## CFAC1_010006100           Previous IDs: CfaC1_KB217577_0120;maker-CfaC1.Contig96-pred_gff_Automagi-gene-0.99
##                                                            annot_gene_product
## CFAC1_010005600  translation initiation factor EIF-2b alpha subunit, putative
## CFAC1_010005700                      PITH domain containing protein, putative
## CFAC1_010005800                           proteasome beta-1 subunit, putative
## CFAC1_010005900  Mitochondrial ribosomal death-associated protein 3, putative
## CFAC1_010006000                               hypothetical protein, conserved
## CFAC1_010006100 Protein phosphatase 2A homologues, catalytic domain, putative
##                 annot_gene_source_id annot_gene_total_hts_snps
## CFAC1_010005600      CFAC1_010005600                      22.0
## CFAC1_010005700      CFAC1_010005700                      23.0
## CFAC1_010005800      CFAC1_010005800                       9.0
## CFAC1_010005900      CFAC1_010005900                      44.0
## CFAC1_010006000      CFAC1_010006000                       9.0
## CFAC1_010006100      CFAC1_010006100                      21.0
##                 annot_gene_transcript_count annot_gene_type
## CFAC1_010005600                         1.0  protein coding
## CFAC1_010005700                         1.0  protein coding
## CFAC1_010005800                         1.0  protein coding
## CFAC1_010005900                         1.0  protein coding
## CFAC1_010006000                         1.0  protein coding
## CFAC1_010006100                         1.0  protein coding
##                 annot_has_missing_transcripts annot_interpro_description
## CFAC1_010005600                            no                        N/A
## CFAC1_010005700                            no                        N/A
## CFAC1_010005800                            no                        N/A
## CFAC1_010005900                            no                        N/A
## CFAC1_010006000                            no                        N/A
## CFAC1_010006100                            no                        N/A
##                 annot_interpro_id annot_is_pseudo annot_isoelectric_point
## CFAC1_010005600               N/A              No                    5.83
## CFAC1_010005700               N/A              No                    4.52
## CFAC1_010005800               N/A              No                    7.69
## CFAC1_010005900               N/A              No                    8.64
## CFAC1_010006000               N/A              No                    9.41
## CFAC1_010006100               N/A              No                    6.15
##                      annot_location_text annot_matched_result
## CFAC1_010005600   CfaC1_01:8824..9729(+)                    Y
## CFAC1_010005700 CfaC1_01:10274..10921(+)                    Y
## CFAC1_010005800 CfaC1_01:11550..12398(+)                    Y
## CFAC1_010005900 CfaC1_01:13166..14665(+)                    Y
## CFAC1_010006000 CfaC1_01:15474..15797(+)                    Y
## CFAC1_010006100 CfaC1_01:16133..17392(+)                    Y
##                 annot_molecular_weight              annot_organism
## CFAC1_010005600                33224.0 C. fasciculata strain Cf-Cl
## CFAC1_010005700                23417.0 C. fasciculata strain Cf-Cl
## CFAC1_010005800                30147.0 C. fasciculata strain Cf-Cl
## CFAC1_010005900                57198.0 C. fasciculata strain Cf-Cl
## CFAC1_010006000                11902.0 C. fasciculata strain Cf-Cl
## CFAC1_010006100                47219.0 C. fasciculata strain Cf-Cl
##                                             annot_pfam_description
## CFAC1_010005600                 Initiation factor 2 subunit family
## CFAC1_010005700                                        PITH domain
## CFAC1_010005800                                 Proteasome subunit
## CFAC1_010005900 Mitochondrial ribosomal death-associated protein 3
## CFAC1_010006000                                                N/A
## CFAC1_010006100                                                N/A
##                 annot_pfam_id annot_pirsf_description annot_pirsf_id
## CFAC1_010005600       PF01008                     N/A            N/A
## CFAC1_010005700       PF06201                     N/A            N/A
## CFAC1_010005800       PF00227                     N/A            N/A
## CFAC1_010005900       PF10236                     N/A            N/A
## CFAC1_010006000           N/A                     N/A            N/A
## CFAC1_010006100           N/A                     N/A            N/A
##                 annot_predicted_go_component
## CFAC1_010005600                             
## CFAC1_010005700                             
## CFAC1_010005800      proteasome core complex
## CFAC1_010005900                             
## CFAC1_010006000                             
## CFAC1_010006100                             
##                           annot_predicted_go_function
## CFAC1_010005600                                      
## CFAC1_010005700                                      
## CFAC1_010005800 threonine-type endopeptidase activity
## CFAC1_010005900                                      
## CFAC1_010006000                                      
## CFAC1_010006100                    hydrolase activity
##                 annot_predicted_go_id_component
## CFAC1_010005600                                
## CFAC1_010005700                                
## CFAC1_010005800                      GO:0005839
## CFAC1_010005900                                
## CFAC1_010006000                                
## CFAC1_010006100                                
##                 annot_predicted_go_id_function
## CFAC1_010005600                               
## CFAC1_010005700                               
## CFAC1_010005800                     GO:0004298
## CFAC1_010005900                               
## CFAC1_010006000                               
## CFAC1_010006100                     GO:0016787
##                 annot_predicted_go_id_process
## CFAC1_010005600                    GO:0044237
## CFAC1_010005700                              
## CFAC1_010005800                    GO:0051603
## CFAC1_010005900                              
## CFAC1_010006000                              
## CFAC1_010006100                              
##                                                 annot_predicted_go_process
## CFAC1_010005600                                 cellular metabolic process
## CFAC1_010005700                                                           
## CFAC1_010005800 proteolysis involved in cellular protein catabolic process
## CFAC1_010005900                                                           
## CFAC1_010006000                                                           
## CFAC1_010006100                                                           
##                 annot_project_id    annot_prositeprofiles_description
## CFAC1_010005600        TriTrypDB                                  N/A
## CFAC1_010005700        TriTrypDB                  PITH domain profile
## CFAC1_010005800        TriTrypDB Proteasome beta-type subunit profile
## CFAC1_010005900        TriTrypDB                                  N/A
## CFAC1_010006000        TriTrypDB                                  N/A
## CFAC1_010006100        TriTrypDB                                  N/A
##                 annot_prositeprofiles_id annot_protein_length
## CFAC1_010005600                      N/A                  301
## CFAC1_010005700                  PS51532                  215
## CFAC1_010005800                  PS51476                  282
## CFAC1_010005900                      N/A                  499
## CFAC1_010006000                      N/A                  107
## CFAC1_010006100                      N/A                  419
##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              annot_protein_sequence
## CFAC1_010005600                                                                                                                                                                                                       MDVPSEDYDVSMMIKAINEVQAILTSRPFATFSEVDTEIAKTLKRYEAFGDGFERFHVNLTKDALQSNDLQKSLKDMNKRCQERLRDCECSQKDQINDILPFVRSTSSILVHGSGNLLALAIACAIQEHEGVRFYVCEGRPARKGFPNGSGEKLLEKVAATPEGMRIKDRLHKYCTIVPDSGVSSVMNAVDFVVMGANCVTEHGGLVHSTGSLQIAIVAAFRNVPFYVLCETFKFAHIFPLGTEDLKQPEDGAGQVVPLVEFVPPSMIALIFSDQGIMPPSAVADEMFRFHTAPSAPWKHP
## CFAC1_010005700                                                                                                                                                                                                                                                                                             MPCNCRQGTALHNHLGDQLSSEGVFGIGVSINEAVDLAALQVWNCKNSAQEAGRLFHPNLGDNTLPICSDDDQELLIFAPLSSVCRVKGLSIKAPASECAPSRVKIFVNIVDITGFGSVQRLVPQEQLQLADGGSEDRIVYRVNPGKFSSVSSLTFFFDESFNGDETDVLRIELFGENTGKSTQQQVATNIVYEGRGNPADHRTAEEKEALFEVR
## CFAC1_010005800                                                                                                                                                                                                                          MLQRPDHTLLQEPSYPKDVQAKLTENGPAQAGKQLFQADPAVIDPQLSEAVSLGTTILAVSYKGGVVLAADSRTSTGAYVVNRVSNKLTKLTKKIYCCRSGSAADTQALAERVSNYLGSYQTDVGADVNVATAANLFQKMCYMNKWNISAGIIVAGYDPINGGSVYSIPSGGSCVKLDYALGGSGSIFLYSFFDANYKPGMTKSECVAFCQRAVAHAYSRDGSSGGLIRTIALDADEPEDQTVPWNKTPYCMEKDPKYQEQAVQNPPFSSSAKITANRTSSA
## CFAC1_010005900 MKGQLVRKGVAALYTAPKIASYTDACEAPTPMWEGAIPLKEERIFKNGVATRIVKKTYSHPPEIPPVHLGFNDVSSMYCLGNDELMKYFPEGLGGKVMQMMPPGHPRGFLYRAQSHLLNCFIDKLPFWQAKEKALRSLMNGRSGFILDGPTGTGKSAIMCQAVHYARSRDILTLYIPNAKAWTHGEWCWPSTILPGFFDAPDAAREFLRTTAEANRSLLEGWDLKVTPQDLPVEQGEKLPRTLYDLCEWGHRAVAPASIDRQSVCIKFFLDEISAEKTKPIVIAVDGWNLFSHETHFRYPHPDFLRTLTTLNDGSTDVDLYPQELPRIPASRLSFVRGLNKLILSKNDPNKFFITCTTRDFKPFDGVSGFADVENDRFKNSLDEYAPYDAEKDSFFHPIKVDNFSDYEYRSFLRFAINSGELAGLGWGPMWHYSSDFERKLYKIGFMSDLNPQRVVDHYHQELVWRYEYKRTRQKQYLLARRRMWQDAKSSTGGRAAKQ
## CFAC1_010006000                                                                                                                                                                                                                                                                                                                                                                                                         MAKREKKRTALVHPNFDSIKRFYAVAPFSAAERAALPPVYAGAYSRRAAMNAYHVVGSSRDQIVSDAMRTFTPASCTKEVMDLYLEVAALVSLFYDTKEQLLALPDI
## CFAC1_010006100                                                                                 MSDAERDLVVWRAAEHIRSGTCVELPRVVRTIRSATEHLKAQQTKLVERVSGSTVVFGPLRGHACDFANVLLTKVLPNRQVTNLVFLGNYIDGAHQSLEVLFLLASLIVCSRFRVTPLIGKHEMMYPLQPENFGSLRNEMLLFCARSPHTSLDDCERAMRDFFAVLPIACVVDNKYFCVAGGPASGFRTLEEMEGKRVTQDALKEFVLNEPMDEDEERIADGSAFVASQSNELAFRFTFNAACNFLSRNRLAMFVVGMEYHTSRPDYDSFARPNHYKESIYFPGYMLGRIHPDTQVPAVLTMFSAPSFCGVNRNNACIAEITDRRLEVQELSVYADRPLVTPGTQDHAFSWAQLMLERAVVAIAREILFGAIKEDAPPESDTHHKHLEEIAVAKMRRMCQLLQTHDLPLPEVPKLPVDE
##                 annot_sequence_id annot_signalp_peptide
## CFAC1_010005600          CfaC1_01                      
## CFAC1_010005700          CfaC1_01                      
## CFAC1_010005800          CfaC1_01                      
## CFAC1_010005900          CfaC1_01                      
## CFAC1_010006000          CfaC1_01                      
## CFAC1_010006100          CfaC1_01                      
##                 annot_signalp_scores annot_smart_description
## CFAC1_010005600                                          N/A
## CFAC1_010005700                                          N/A
## CFAC1_010005800                                          N/A
## CFAC1_010005900                                          N/A
## CFAC1_010006000                                          N/A
## CFAC1_010006100                                             
##                 annot_smart_id   annot_source_id annot_strand
## CFAC1_010005600            N/A CFAC1_010005600.1      forward
## CFAC1_010005700            N/A CFAC1_010005700.1      forward
## CFAC1_010005800            N/A CFAC1_010005800.1      forward
## CFAC1_010005900            N/A CFAC1_010005900.1      forward
## CFAC1_010006000            N/A CFAC1_010006000.1      forward
## CFAC1_010006100        SM00156 CFAC1_010006100.1      forward
##                                           annot_superfamily_description
## CFAC1_010005600                          NagB/RpiA/CoA transferase-like
## CFAC1_010005700                           Galactose-binding domain-like
## CFAC1_010005800 N-terminal nucleophile aminohydrolases (Ntn hydrolases)
## CFAC1_010005900    P-loop containing nucleoside triphosphate hydrolases
## CFAC1_010006000                                                     N/A
## CFAC1_010006100                          Metallo-dependent phosphatases
##                 annot_superfamily_id annot_three_prime_utr_length
## CFAC1_010005600            SSF100950                             
## CFAC1_010005700             SSF49785                             
## CFAC1_010005800             SSF56235                             
## CFAC1_010005900             SSF52540                             
## CFAC1_010006000                  N/A                             
## CFAC1_010006100             SSF56300                             
##                 annot_tigrfam_description annot_tigrfam_id annot_tm_count
## CFAC1_010005600                       N/A              N/A              0
## CFAC1_010005700                       N/A              N/A              0
## CFAC1_010005800                       N/A              N/A              0
## CFAC1_010005900                       N/A              N/A              0
## CFAC1_010006000                       N/A              N/A              0
## CFAC1_010006100                       N/A              N/A              0
##                 annot_trans_found_per_gene_internal
## CFAC1_010005600                                   1
## CFAC1_010005700                                   1
## CFAC1_010005800                                   1
## CFAC1_010005900                                   1
## CFAC1_010006000                                   1
## CFAC1_010006100                                   1
##                 annot_transcript_index_per_gene annot_transcript_length
## CFAC1_010005600                               1                   906.0
## CFAC1_010005700                               1                   648.0
## CFAC1_010005800                               1                   849.0
## CFAC1_010005900                               1                  1500.0
## CFAC1_010006000                               1                   324.0
## CFAC1_010006100                               1                  1260.0
##                 annot_transcript_link
## CFAC1_010005600     CFAC1_010005600.1
## CFAC1_010005700     CFAC1_010005700.1
## CFAC1_010005800     CFAC1_010005800.1
## CFAC1_010005900     CFAC1_010005900.1
## CFAC1_010006000     CFAC1_010006000.1
## CFAC1_010006100     CFAC1_010006100.1
##                                                      annot_transcript_product
## CFAC1_010005600  translation initiation factor EIF-2b alpha subunit, putative
## CFAC1_010005700                      PITH domain containing protein, putative
## CFAC1_010005800                           proteasome beta-1 subunit, putative
## CFAC1_010005900  Mitochondrial ribosomal death-associated protein 3, putative
## CFAC1_010006000                               hypothetical protein, conserved
## CFAC1_010006100 Protein phosphatase 2A homologues, catalytic domain, putative
##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    annot_transcript_sequence
## CFAC1_010005600                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   ATGGACGTGCCAAGCGAAGATTATGACGTGAGTATGATGATAAAAGCCATCAACGAGGTGCAGGCAATTCTCACCTCGCGTCCGTTTGCGACGTTCTCGGAAGTCGATACCGAGATCGCCAAGACGCTGAAACGGTACGAGGCGTTTGGCGACGGCTTTGAGCGCTTTCACGTGAACCTCACAAAAGACGCACTCCAGTCGAATGACTTACAGAAGAGTTTGAAAGACATGAACAAGCGCTGCCAAGAGCGGCTGCGAGACTGTGAATGCAGTCAAAAGGACCAAATTAACGATATCCTTCCTTTTGTTCGCAGTACGTCTTCCATTCTCGTCCACGGCTCTGGCAACCTTTTAGCGCTTGCTATTGCTTGTGCCATTCAGGAGCACGAGGGCGTTCGCTTCTACGTTTGTGAAGGCCGCCCGGCACGCAAAGGCTTCCCGAACGGCTCAGGTGAGAAGCTGCTAGAGAAGGTCGCTGCGACGCCAGAGGGAATGCGTATCAAGGACAGGCTTCACAAGTACTGCACAATTGTTCCAGACAGTGGTGTCAGCTCCGTGATGAATGCAGTTGACTTTGTAGTAATGGGAGCCAACTGTGTCACTGAGCACGGTGGTCTTGTGCACTCCACCGGGTCTCTGCAGATCGCCATCGTTGCCGCGTTTCGCAATGTGCCCTTTTATGTTTTGTGTGAAACCTTCAAGTTTGCGCACATCTTTCCTCTGGGCACAGAGGACCTCAAGCAGCCGGAGGACGGTGCTGGACAGGTTGTCCCGCTTGTGGAGTTTGTCCCTCCATCGATGATCGCTTTGATTTTCTCCGACCAGGGAATTATGCCTCCGTCAGCTGTGGCGGACGAGATGTTTCGCTTCCACACCGCGCCGTCTGCACCATGGAAGCACCCGTAA
## CFAC1_010005700                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     ATGCCGTGCAACTGCCGTCAAGGGACTGCGCTGCACAACCACCTTGGTGACCAGCTCTCCAGCGAAGGCGTGTTTGGCATCGGAGTTTCGATCAACGAAGCAGTCGACCTCGCGGCACTGCAGGTGTGGAACTGCAAGAACAGTGCCCAGGAGGCGGGTCGCCTATTCCATCCGAACCTGGGCGACAACACGTTGCCCATTTGCAGCGATGACGACCAGGAGCTGCTGATTTTCGCACCACTGAGTTCCGTATGTCGTGTCAAGGGCCTGAGCATCAAGGCACCTGCGAGCGAGTGTGCGCCGTCTCGCGTAAAGATCTTTGTGAACATTGTGGACATCACGGGTTTCGGCTCTGTCCAGAGACTGGTGCCACAAGAACAGCTGCAGCTTGCTGACGGCGGCTCAGAAGACCGGATCGTGTACCGCGTGAATCCCGGCAAGTTCTCCTCCGTCTCGTCACTCACGTTCTTCTTCGATGAGAGCTTCAACGGCGACGAGACAGATGTGCTGCGCATCGAGCTCTTTGGAGAGAACACTGGCAAGTCGACGCAGCAGCAGGTCGCCACGAACATTGTGTATGAAGGTCGCGGAAACCCTGCGGACCATCGTACTGCGGAGGAGAAGGAGGCTCTTTTTGAGGTGCGGTGA
## CFAC1_010005800                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            ATGCTCCAGCGTCCCGACCACACTCTCCTGCAGGAGCCGTCGTATCCGAAGGATGTCCAAGCCAAGCTGACGGAGAACGGACCGGCGCAGGCGGGCAAGCAGCTCTTCCAGGCCGACCCTGCCGTCATCGACCCGCAGCTGAGCGAGGCGGTCTCGCTGGGCACCACCATTCTTGCGGTGTCGTACAAGGGCGGTGTCGTGCTCGCCGCTGACTCGCGCACGTCCACTGGTGCGTACGTGGTGAACCGTGTCAGCAACAAGCTCACCAAGCTGACAAAGAAGATCTACTGCTGCCGTAGCGGGTCTGCTGCCGACACGCAGGCCCTAGCGGAGCGCGTCTCCAACTACCTTGGTAGCTACCAGACAGATGTCGGTGCCGACGTCAACGTCGCCACCGCTGCCAACCTGTTCCAGAAGATGTGCTACATGAACAAGTGGAACATCTCCGCTGGTATCATCGTCGCCGGCTACGATCCCATCAACGGCGGCTCTGTCTACAGCATTCCGTCCGGCGGCTCGTGTGTGAAGCTGGACTATGCGCTCGGTGGCAGCGGCTCCATTTTCCTGTACTCGTTCTTCGATGCCAACTATAAGCCTGGCATGACGAAGAGCGAGTGCGTCGCGTTCTGCCAGAGGGCAGTGGCGCACGCGTACAGCCGCGACGGGTCCAGCGGTGGTCTGATCCGGACCATCGCACTCGACGCCGACGAGCCGGAGGACCAGACGGTGCCGTGGAACAAGACGCCGTACTGCATGGAGAAGGACCCCAAGTACCAGGAGCAGGCGGTCCAGAACCCTCCGTTCAGCAGCAGCGCGAAGATTACGGCAAACCGTACGTCGTCGGCGTGA
## CFAC1_010005900 ATGAAGGGGCAGCTGGTCCGGAAAGGCGTAGCGGCGTTGTACACGGCGCCGAAGATTGCGTCGTACACAGATGCTTGCGAGGCACCAACGCCAATGTGGGAGGGTGCGATTCCGCTGAAAGAAGAGCGCATTTTCAAAAACGGTGTTGCGACGCGGATCGTGAAGAAGACGTACTCGCACCCGCCGGAGATTCCACCGGTGCACTTAGGCTTCAACGACGTCAGCTCCATGTACTGTCTCGGCAACGACGAGCTCATGAAGTACTTCCCCGAGGGTCTTGGCGGGAAGGTGATGCAAATGATGCCGCCAGGACACCCGCGCGGCTTCCTGTACCGCGCGCAGAGCCACCTGCTCAACTGCTTCATCGACAAGTTGCCCTTCTGGCAAGCAAAGGAGAAGGCACTCCGCTCCTTAATGAACGGGCGGAGCGGCTTTATCCTTGACGGACCAACCGGCACTGGAAAGAGCGCCATCATGTGTCAAGCAGTCCACTACGCGCGCTCGCGCGACATCCTGACGCTCTACATCCCAAATGCGAAGGCGTGGACGCACGGCGAGTGGTGCTGGCCAAGCACCATCCTTCCTGGCTTTTTCGACGCGCCTGACGCCGCGCGTGAGTTTCTGCGCACGACCGCAGAGGCGAATCGCAGCCTGCTGGAGGGGTGGGACCTGAAGGTGACGCCGCAGGATCTGCCCGTCGAGCAAGGTGAGAAGCTGCCACGGACGCTCTACGACCTGTGCGAGTGGGGCCACCGCGCCGTCGCTCCAGCCTCGATCGATCGTCAGAGCGTGTGCATCAAGTTTTTCCTTGACGAGATTAGCGCCGAAAAGACAAAGCCGATTGTGATCGCGGTGGACGGGTGGAACCTCTTCTCGCACGAGACGCACTTCCGTTACCCGCACCCCGACTTCCTGCGCACGCTCACGACGCTCAACGATGGGTCGACGGACGTCGACCTCTACCCGCAGGAGCTGCCGCGCATTCCGGCGTCGCGTCTCAGCTTCGTGCGCGGGCTCAACAAGCTCATCCTCTCCAAGAACGACCCAAACAAGTTTTTCATTACCTGCACGACGCGCGACTTTAAGCCCTTCGACGGCGTCAGCGGCTTCGCGGATGTGGAGAACGACAGGTTCAAAAACAGCCTCGACGAGTATGCACCGTACGACGCCGAGAAGGACTCCTTCTTCCACCCGATCAAAGTCGACAACTTCAGCGACTACGAGTACCGCTCCTTCCTGCGTTTTGCCATCAACTCGGGCGAACTGGCAGGACTCGGGTGGGGGCCGATGTGGCACTACTCAAGCGACTTTGAGCGGAAGCTGTACAAGATCGGCTTTATGTCGGATCTCAACCCGCAGCGCGTCGTCGATCACTACCACCAGGAGCTTGTGTGGCGCTACGAGTACAAGCGCACCCGCCAGAAGCAGTACCTGCTGGCGCGCCGCAGAATGTGGCAGGATGCGAAGTCGTCGACCGGCGGTCGCGCGGCGAAGCAGTGA
## CFAC1_010006000                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         ATGGCCAAGCGAGAGAAGAAGCGAACGGCTTTGGTGCATCCCAACTTCGACAGTATCAAGCGCTTCTACGCCGTCGCGCCTTTTAGCGCCGCAGAGCGCGCCGCGCTGCCGCCGGTGTACGCCGGTGCCTATAGCCGACGCGCCGCGATGAACGCCTACCACGTCGTTGGCTCGTCACGAGACCAGATCGTGTCAGATGCCATGCGCACTTTCACCCCGGCGTCGTGCACAAAGGAGGTGATGGATCTGTACTTGGAAGTGGCGGCGCTAGTCTCACTGTTTTACGATACGAAGGAGCAGCTCCTTGCGCTGCCGGACATCTAA
## CFAC1_010006100                                                                                                                                                                                                                                                 ATGAGCGATGCGGAGAGAGATCTGGTGGTGTGGAGGGCAGCGGAGCACATCCGGTCCGGCACCTGCGTGGAGCTGCCGCGTGTGGTGCGGACGATCCGCTCTGCAACAGAACACCTGAAGGCGCAGCAAACGAAGTTGGTGGAGAGGGTTTCGGGAAGCACCGTCGTCTTCGGTCCTCTCCGCGGTCACGCCTGCGACTTCGCCAACGTTCTGCTTACCAAGGTCCTGCCCAACCGACAGGTGACAAATCTCGTCTTTCTCGGCAACTACATCGACGGCGCTCACCAGTCTCTCGAGGTGCTCTTCCTCCTTGCCTCTCTCATCGTCTGCTCTCGATTCCGCGTCACTCCGCTTATCGGCAAGCATGAGATGATGTACCCTCTCCAGCCGGAGAACTTCGGCTCGCTGCGCAACGAAATGCTCCTCTTCTGTGCCCGCTCGCCGCACACGTCCCTCGACGACTGCGAGCGCGCCATGCGGGACTTCTTTGCCGTCCTGCCGATCGCCTGTGTGGTGGACAACAAGTACTTCTGTGTAGCTGGAGGACCGGCGTCCGGCTTCCGCACACTAGAGGAGATGGAGGGGAAGCGGGTGACGCAGGACGCGCTGAAGGAGTTCGTGCTGAACGAGCCGATGGACGAGGACGAGGAGCGCATTGCGGACGGCAGTGCTTTTGTGGCGTCGCAGAGCAACGAGCTGGCCTTTCGCTTCACCTTCAACGCGGCGTGCAACTTCTTGAGCCGCAACCGCCTGGCGATGTTTGTGGTAGGCATGGAGTACCACACGAGCCGACCCGATTACGACAGCTTCGCCCGCCCGAACCACTACAAGGAGTCCATCTACTTTCCAGGCTACATGCTGGGCCGCATCCATCCCGACACGCAGGTGCCGGCGGTGCTGACCATGTTCTCCGCACCGTCCTTCTGCGGCGTCAACCGCAACAACGCCTGCATTGCTGAGATCACCGACCGGCGGCTGGAGGTGCAGGAGCTGAGTGTCTACGCTGACCGCCCGCTTGTCACGCCAGGCACGCAGGACCACGCCTTTAGCTGGGCACAGCTGATGCTTGAGCGCGCCGTTGTGGCCATTGCCCGAGAGATTTTGTTTGGTGCCATCAAGGAGGATGCGCCGCCGGAGAGCGATACCCACCACAAGCACCTGGAGGAGATCGCGGTGGCGAAGATGCGGCGCATGTGTCAGCTACTGCAGACGCACGATCTGCCCTTGCCGGAGGTGCCGAAGCTCCCTGTCGACGAGTGA
##                 annot_transcripts_found_per_gene annot_uniprot_id
## CFAC1_010005600                           1 of 1                 
## CFAC1_010005700                           1 of 1                 
## CFAC1_010005800                           1 of 1                 
## CFAC1_010005900                           1 of 1                 
## CFAC1_010006000                           1 of 1                 
## CFAC1_010006100                           1 of 1                 
##                 annot_uri annot_wdk_weight
## CFAC1_010005600                          0
## CFAC1_010005700                          0
## CFAC1_010005800                          0
## CFAC1_010005900                          0
## CFAC1_010006000                          0
## CFAC1_010006100                          0</code></pre>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb116-1" title="1"><span class="co">## Yay! We have data about Crithidia!</span></a>
<a class="sourceLine" id="cb116-2" title="2"></a>
<a class="sourceLine" id="cb116-3" title="3">chosen_columns_idx &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="dt">x=</span>avail_columns, <span class="dt">pattern=</span><span class="st">&quot;^GO&quot;</span>)</a>
<a class="sourceLine" id="cb116-4" title="4">chosen_columns &lt;-<span class="st"> </span>avail_columns[chosen_columns_idx]</a>
<a class="sourceLine" id="cb116-5" title="5">crit_go &lt;-<span class="st"> </span><span class="kw">load_orgdb_go</span>(orgdb_pkg, <span class="dt">keytype=</span><span class="st">&quot;gid&quot;</span>, <span class="dt">columns=</span>chosen_columns)</a></code></pre></div>
<pre><code>## Loading required namespace: GO.db</code></pre>
<pre><code>## </code></pre>
<pre><code>## This is an orgdb, good.</code></pre>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb121-1" title="1"><span class="kw">head</span>(crit_go, <span class="dt">n=</span><span class="dv">1</span>)</a></code></pre></div>
<pre><code>## # A tibble: 1 x 11
##   GID   GO_EVIDENCE_CODE GO_GO_ID GO_GO_TERM_NAME GO_IS_NOT GO_ONTOLOGY
##   &lt;chr&gt; &lt;chr&gt;            &lt;chr&gt;    &lt;chr&gt;           &lt;chr&gt;     &lt;chr&gt;      
## 1 CFAC… IEA              GO:0044… cellular metab… N/A       Biological…
## # … with 5 more variables: GO_REFERENCE &lt;chr&gt;, GO_SORT_KEY &lt;chr&gt;,
## #   GO_SOURCE &lt;chr&gt;, GO_SUPPORT_FOR_EVIDENCE_CODE_ASSIGNMENT &lt;chr&gt;,
## #   GO_TRANSCRIPT_ID_S_ &lt;chr&gt;</code></pre>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb123-1" title="1"><span class="co">## Yay Gene ontology data for Crithidia!</span></a>
<a class="sourceLine" id="cb123-2" title="2"></a>
<a class="sourceLine" id="cb123-3" title="3">chosen_columns_idx &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="dt">x=</span>avail_columns, <span class="dt">pattern=</span><span class="st">&quot;^INTERPRO&quot;</span>)</a>
<a class="sourceLine" id="cb123-4" title="4">chosen_columns &lt;-<span class="st"> </span>avail_columns[chosen_columns_idx]</a>
<a class="sourceLine" id="cb123-5" title="5">crit_interpro &lt;-<span class="st"> </span><span class="kw">load_orgdb_go</span>(orgdb_pkg, <span class="dt">keytype=</span><span class="st">&quot;gid&quot;</span>, <span class="dt">columns=</span>chosen_columns)</a></code></pre></div>
<pre><code>## This is an orgdb, good.
## 'select()' returned 1:many mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb125-1" title="1"><span class="kw">head</span>(crit_interpro)</a></code></pre></div>
<pre><code>## # A tibble: 6 x 10
##   GID   INTERPRO_DESCRI… INTERPRO_E_VALUE INTERPRO_END_MIN INTERPRO_INTERP…
##   &lt;chr&gt; &lt;chr&gt;            &lt;chr&gt;            &lt;chr&gt;            &lt;chr&gt;           
## 1 CFAC… Initiation fact… 2.1e-17          280              IPR000649       
## 2 CFAC… NagB/RpiA/CoA t… 4.2e-47          290              IPR037171       
## 3 CFAC… PITH domain      4.7e-24          177              IPR010400       
## 4 CFAC… PITH domain pro… 37.0             195              IPR010400       
## 5 CFAC… Galactose-bindi… 1.2e-30          177              IPR008979       
## 6 CFAC… Proteasome subu… 2.9e-41          228              IPR001353       
## # … with 5 more variables: INTERPRO_NAME &lt;chr&gt;, INTERPRO_PRIMARY_ID &lt;chr&gt;,
## #   INTERPRO_SECONDARY_ID &lt;chr&gt;, INTERPRO_START_MIN &lt;chr&gt;,
## #   INTERPRO_TRANSCRIPT_ID_S_ &lt;chr&gt;</code></pre>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb127-1" title="1"><span class="co">## Interpro data for Crithidia!</span></a>
<a class="sourceLine" id="cb127-2" title="2"></a>
<a class="sourceLine" id="cb127-3" title="3">chosen_columns_idx &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="dt">x=</span>avail_columns, <span class="dt">pattern=</span><span class="st">&quot;^PATHWAY&quot;</span>)</a>
<a class="sourceLine" id="cb127-4" title="4">chosen_columns &lt;-<span class="st"> </span>avail_columns[chosen_columns_idx]</a>
<a class="sourceLine" id="cb127-5" title="5">crit_path &lt;-<span class="st"> </span><span class="kw">load_orgdb_go</span>(orgdb_pkg, <span class="dt">keytype=</span><span class="st">&quot;gid&quot;</span>, <span class="dt">columns=</span>chosen_columns)</a></code></pre></div>
<pre><code>## This is an orgdb, good.
## 'select()' returned 1:many mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb129-1" title="1"><span class="kw">head</span>(crit_path)</a></code></pre></div>
<pre><code>## # A tibble: 6 x 8
##   GID   PATHWAY_EC_NUMB… PATHWAY_EXACT_E… PATHWAY_EXPASY_… PATHWAY_PATHWAY
##   &lt;chr&gt; &lt;chr&gt;            &lt;chr&gt;            &lt;chr&gt;            &lt;chr&gt;          
## 1 CFAC… 5.3.1.23         Yes              http://enzyme.e… &lt;i&gt;S&lt;/i&gt;-methy…
## 2 CFAC… 5.-.-.-          No               http://enzyme.e… hypoglycin bio…
## 3 CFAC… 5.3.1.-          No               http://enzyme.e… fluoroacetate …
## 4 CFAC… 5.3.1.23         Yes              http://enzyme.e… &lt;i&gt;S&lt;/i&gt;-methy…
## 5 CFAC… 5.3.1.-          No               http://enzyme.e… dTDP-L-megosam…
## 6 CFAC… 5.3.1.-          No               http://enzyme.e… superpathway o…
## # … with 3 more variables: PATHWAY_PATHWAY_SOURCE &lt;chr&gt;,
## #   PATHWAY_PATHWAY_SOURCE_ID &lt;chr&gt;,
## #   PATHWAY_REACTIONS_MATCHING_EC_NUMBER &lt;chr&gt;</code></pre>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb131-1" title="1"><span class="co">## Pathway data for Crithidia!</span></a>
<a class="sourceLine" id="cb131-2" title="2"></a>
<a class="sourceLine" id="cb131-3" title="3">chosen_columns_idx &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="dt">x=</span>avail_columns, <span class="dt">pattern=</span><span class="st">&quot;^ORTHOLOG&quot;</span>)</a>
<a class="sourceLine" id="cb131-4" title="4">chosen_columns &lt;-<span class="st"> </span>avail_columns[chosen_columns_idx]</a>
<a class="sourceLine" id="cb131-5" title="5">chosen_columns &lt;-<span class="st"> </span><span class="kw">c</span>(chosen_columns, <span class="st">&quot;ORGANISM&quot;</span>)</a>
<a class="sourceLine" id="cb131-6" title="6">crit_ortho &lt;-<span class="st"> </span><span class="kw">load_orgdb_go</span>(orgdb_pkg, <span class="dt">keytype=</span><span class="st">&quot;gid&quot;</span>, <span class="dt">columns=</span>chosen_columns)</a></code></pre></div>
<pre><code>## This is an orgdb, good.
## 'select()' returned 1:many mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb133-1" title="1"><span class="kw">head</span>(crit_ortho)</a></code></pre></div>
<pre><code>## # A tibble: 6 x 5
##   GID     ORTHOLOG_COUNT ORTHOLOG_GROUP            ORTHOLOG_ID  ORGANISM   
##   &lt;chr&gt;   &lt;chr&gt;          &lt;chr&gt;                     &lt;chr&gt;        &lt;chr&gt;      
## 1 CFAC1_… 39             &lt;a href=http://orthomcl.… BCY84_22384  T. cruzi c…
## 2 CFAC1_… 39             &lt;a href=http://orthomcl.… Baya_067_03… B. ayalai …
## 3 CFAC1_… 39             &lt;a href=http://orthomcl.… CFAC1_01000… C. fascicu…
## 4 CFAC1_… 39             &lt;a href=http://orthomcl.… DQ04_017610… T. grayi A…
## 5 CFAC1_… 39             &lt;a href=http://orthomcl.… EMOLV88_120… E. montero…
## 6 CFAC1_… 39             &lt;a href=http://orthomcl.… LAEL147_000… L. aethiop…</code></pre>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb135-1" title="1"><span class="co">## Orthologs!</span></a></code></pre></div>
</div>
<div id="shortcut-methods" class="section level3">
<h3>Shortcut methods</h3>
<p>Remembering the keytype and package names and everything can be annoying. The following attempts to make that easier. In this instance, the only thing we need to provide is a unique substring for the species of interest. If the substring is not unique, this should show the matches and choose the first.</p>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb136-1" title="1"><span class="co">## The function load_eupath_annotations() provides a shortcut to the above.</span></a>
<a class="sourceLine" id="cb136-2" title="2">crit_annotv2 &lt;-<span class="st"> </span><span class="kw">load_eupath_annotations</span>(<span class="dt">species=</span><span class="st">&quot;Crith&quot;</span>)</a></code></pre></div>
<pre><code>## Found the following hits: Crithidia fasciculata strain Cf-Cl, choosing the first.</code></pre>
<pre><code>## Unable to find CDSNAME, setting it to ANNOT_GENE_NAME.</code></pre>
<pre><code>## Unable to find CDSCHROM in the db, removing it.</code></pre>
<pre><code>## Unable to find CDSSTRAND in the db, removing it.</code></pre>
<pre><code>## Unable to find CDSSTART in the db, removing it.</code></pre>
<pre><code>## Unable to find CDSEND in the db, removing it.</code></pre>
<pre><code>## Extracted all gene ids.</code></pre>
<pre><code>## Attempting to select: ANNOT_GENE_NAME, GENE_TYPE, GID, ANNOT_ANNOTATED_GO_COMPONENT, ANNOT_ANNOTATED_GO_FUNCTION, ANNOT_ANNOTATED_GO_ID_COMPONENT, ANNOT_ANNOTATED_GO_ID_FUNCTION, ANNOT_ANNOTATED_GO_ID_PROCESS, ANNOT_ANNOTATED_GO_PROCESS, ANNOT_CDS, ANNOT_CDS_LENGTH, ANNOT_CHROMOSOME, ANNOT_EC_NUMBERS, ANNOT_EC_NUMBERS_DERIVED, ANNOT_EXON_COUNT, ANNOT_FIVE_PRIME_UTR_LENGTH, ANNOT_GENE_ENTREZ_ID, ANNOT_GENE_EXON_COUNT, ANNOT_GENE_HTS_NONCODING_SNPS, ANNOT_GENE_HTS_NONSYN_SYN_RATIO, ANNOT_GENE_HTS_NONSYNONYMOUS_SNPS, ANNOT_GENE_HTS_STOP_CODON_SNPS, ANNOT_GENE_HTS_SYNONYMOUS_SNPS, ANNOT_GENE_LOCATION_TEXT, ANNOT_GENE_NAME, ANNOT_GENE_ORTHOLOG_NUMBER, ANNOT_GENE_ORTHOMCL_NAME, ANNOT_GENE_PARALOG_NUMBER, ANNOT_GENE_PREVIOUS_IDS, ANNOT_GENE_PRODUCT, ANNOT_GENE_SOURCE_ID, ANNOT_GENE_TOTAL_HTS_SNPS, ANNOT_GENE_TRANSCRIPT_COUNT, ANNOT_GENE_TYPE, ANNOT_HAS_MISSING_TRANSCRIPTS, ANNOT_INTERPRO_DESCRIPTION, ANNOT_INTERPRO_ID, ANNOT_IS_PSEUDO, ANNOT_ISOELECTRIC_POINT, ANNOT_LOCATION_TEXT, ANNOT_MATCHED_RESULT, ANNOT_MOLECULAR_WEIGHT, ANNOT_ORGANISM, ANNOT_PFAM_DESCRIPTION, ANNOT_PFAM_ID, ANNOT_PIRSF_DESCRIPTION, ANNOT_PIRSF_ID, ANNOT_PREDICTED_GO_COMPONENT, ANNOT_PREDICTED_GO_FUNCTION, ANNOT_PREDICTED_GO_ID_COMPONENT, ANNOT_PREDICTED_GO_ID_FUNCTION, ANNOT_PREDICTED_GO_ID_PROCESS, ANNOT_PREDICTED_GO_PROCESS, ANNOT_PROJECT_ID, ANNOT_PROSITEPROFILES_DESCRIPTION, ANNOT_PROSITEPROFILES_ID, ANNOT_PROTEIN_LENGTH, ANNOT_PROTEIN_SEQUENCE, ANNOT_SEQUENCE_ID, ANNOT_SIGNALP_PEPTIDE, ANNOT_SIGNALP_SCORES, ANNOT_SMART_DESCRIPTION, ANNOT_SMART_ID, ANNOT_SOURCE_ID, ANNOT_STRAND, ANNOT_SUPERFAMILY_DESCRIPTION, ANNOT_SUPERFAMILY_ID, ANNOT_THREE_PRIME_UTR_LENGTH, ANNOT_TIGRFAM_DESCRIPTION, ANNOT_TIGRFAM_ID, ANNOT_TM_COUNT, ANNOT_TRANS_FOUND_PER_GENE_INTERNAL, ANNOT_TRANSCRIPT_INDEX_PER_GENE, ANNOT_TRANSCRIPT_LENGTH, ANNOT_TRANSCRIPT_LINK, ANNOT_TRANSCRIPT_PRODUCT, ANNOT_TRANSCRIPT_SEQUENCE, ANNOT_TRANSCRIPTS_FOUND_PER_GENE, ANNOT_UNIPROT_ID, ANNOT_URI, ANNOT_WDK_WEIGHT</code></pre>
<pre><code>## 'select()' returned 1:1 mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb146-1" title="1"><span class="kw">dim</span>(crit_annotv2)</a></code></pre></div>
<pre><code>## [1] 9489   79</code></pre>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb148-1" title="1"><span class="co">## This provides the same information as the results of the select up above.</span></a></code></pre></div>
</div>
<div id="shortcut-orthologs" class="section level3">
<h3>Shortcut orthologs</h3>
<p>A task I find myself needing to do fairly often is to get the set of genes orthologous to a given gene. There are lots of methods to handle this question; one of them includes the nice table provided by the eupathdb. Let us query that.</p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb149-1" title="1">crit_ortho &lt;-<span class="st"> </span><span class="kw">extract_eupath_orthologs</span>(<span class="st">&quot;Crith&quot;</span>)</a></code></pre></div>
<pre><code>## Starting metadata download.</code></pre>
<pre><code>## Finished metadata download.</code></pre>
<pre><code>## Loaded: org.Cfasciculata.Cf.Cl.v41.eg.db</code></pre>
<pre><code>## Found all the required columns!</code></pre>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<pre><code>## There are 41 possible species in this group.</code></pre>
<pre><code>## Found species: B. ayalai B08-376</code></pre>
<pre><code>## Found species: C. fasciculata strain Cf-Cl</code></pre>
<pre><code>## Found species: E. monterogeii strain LV88</code></pre>
<pre><code>## Found species: L. aethiopica L147</code></pre>
<pre><code>## Found species: L. arabica strain LEM1108</code></pre>
<pre><code>## Found species: L. braziliensis MHOM/BR/75/M2903</code></pre>
<pre><code>## Found species: L. braziliensis MHOM/BR/75/M2904</code></pre>
<pre><code>## Found species: L. donovani BPK282A1</code></pre>
<pre><code>## Found species: L. enriettii strain LEM3045</code></pre>
<pre><code>## Found species: L. gerbilli strain LEM452</code></pre>
<pre><code>## Found species: L. infantum JPCM5</code></pre>
<pre><code>## Found species: L. major strain Friedlin</code></pre>
<pre><code>## Found species: L. major strain LV39c5</code></pre>
<pre><code>## Found species: L. major strain SD 75.1</code></pre>
<pre><code>## Found species: L. mexicana MHOM/GT/2001/U1103</code></pre>
<pre><code>## Found species: L. panamensis MHOM/COL/81/L13</code></pre>
<pre><code>## Found species: L. panamensis strain MHOM/PA/94/PSC-1</code></pre>
<pre><code>## Found species: L. pyrrhocoris H10</code></pre>
<pre><code>## Found species: L. seymouri ATCC 30220</code></pre>
<pre><code>## Found species: L. sp. MAR LEM2494</code></pre>
<pre><code>## Found species: L. tarentolae Parrot-TarII</code></pre>
<pre><code>## Found species: L. tropica L590</code></pre>
<pre><code>## Found species: L. turanica strain LEM423</code></pre>
<pre><code>## Found species: P. confusum CUL13</code></pre>
<pre><code>## Found species: T. brucei brucei TREU927</code></pre>
<pre><code>## Found species: T. brucei gambiense DAL972</code></pre>
<pre><code>## Found species: T. brucei Lister strain 427</code></pre>
<pre><code>## Found species: T. congolense IL3000</code></pre>
<pre><code>## Found species: T. cruzi CL Brener Esmeraldo-like</code></pre>
<pre><code>## Found species: T. cruzi CL Brener Non-Esmeraldo-like</code></pre>
<pre><code>## Found species: T. cruzi cruzi strain Dm28c</code></pre>
<pre><code>## Found species: T. cruzi Dm28c</code></pre>
<pre><code>## Found species: T. cruzi marinkellei strain B7</code></pre>
<pre><code>## Found species: T. cruzi strain CL Brener</code></pre>
<pre><code>## Found species: T. cruzi Sylvio X10/1</code></pre>
<pre><code>## Found species: T. cruzi Sylvio X10/1-2012</code></pre>
<pre><code>## Found species: T. evansi strain STIB 805</code></pre>
<pre><code>## Found species: T. grayi ANR4</code></pre>
<pre><code>## Found species: T. rangeli SC58</code></pre>
<pre><code>## Found species: T. theileri isolate Edinburgh</code></pre>
<pre><code>## Found species: T. vivax Y486</code></pre>
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb197-1" title="1"><span class="kw">dim</span>(crit_ortho)</a></code></pre></div>
<pre><code>## [1] 459383      8</code></pre>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb199-1" title="1"><span class="kw">head</span>(crit_ortho)</a></code></pre></div>
<pre><code>## # A tibble: 6 x 8
## # Groups:   GID [1]
##   GID   ORTHOLOG_ID ORTHOLOG_SPECIES ORTHOLOG_URL ORTHOLOG_COUNT
##   &lt;chr&gt; &lt;chr&gt;       &lt;fct&gt;            &lt;chr&gt;                 &lt;int&gt;
## 1 CFAC… BCY84_22384 T. cruzi cruzi … &lt;a href=htt…             39
## 2 CFAC… Baya_067_0… B. ayalai B08-3… &lt;a href=htt…             39
## 3 CFAC… CFAC1_0100… C. fasciculata … &lt;a href=htt…             39
## 4 CFAC… DQ04_01761… T. grayi ANR4    &lt;a href=htt…             39
## 5 CFAC… EMOLV88_12… E. monterogeii … &lt;a href=htt…             39
## 6 CFAC… LAEL147_00… L. aethiopica L… &lt;a href=htt…             39
## # … with 3 more variables: ORTHOLOG_GROUP &lt;chr&gt;, QUERIES_IN_GROUP &lt;int&gt;,
## #   GROUP_REPRESENTATION &lt;dbl&gt;</code></pre>
<div class="sourceCode" id="cb201"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb201-1" title="1"><span class="kw">summary</span>(crit_ortho)</a></code></pre></div>
<pre><code>##      GID            ORTHOLOG_ID       
##  Length:459383      Length:459383     
##  Class :character   Class :character  
##  Mode  :character   Mode  :character  
##                                       
##                                       
##                                       
##                                       
##                     ORTHOLOG_SPECIES  ORTHOLOG_URL       ORTHOLOG_COUNT 
##  C. fasciculata strain Cf-Cl: 22727   Length:459383      Min.   :  0.0  
##  T. cruzi Sylvio X10/1      : 21286   Class :character   1st Qu.: 38.0  
##  T. cruzi cruzi strain Dm28c: 18560   Mode  :character   Median : 42.0  
##  L. pyrrhocoris H10         : 17888                      Mean   :122.4  
##  E. monterogeii strain LV88 : 15120                      3rd Qu.:123.0  
##  L. infantum JPCM5          : 12424                      Max.   :759.0  
##  (Other)                    :351378                                     
##  ORTHOLOG_GROUP     QUERIES_IN_GROUP GROUP_REPRESENTATION
##  Length:459383      Min.   :  1.0    Min.   : 0.0000     
##  Class :character   1st Qu.: 39.0    1st Qu.: 0.9268     
##  Mode  :character   Median : 43.0    Median : 1.0244     
##                     Mean   :133.4    Mean   : 2.9864     
##                     3rd Qu.:130.0    3rd Qu.: 3.0000     
##                     Max.   :760.0    Max.   :18.5122     
## </code></pre>
</div>
</div>
</div>
<div id="session-information" class="section level1">
<h1>Session Information</h1>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb203-1" title="1"><span class="kw">sessionInfo</span>()</a></code></pre></div>
<pre><code>## R version 3.5.2 (2018-12-20)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Devuan GNU/Linux beowulf/ceres
## 
## Matrix products: default
## BLAS: /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
## LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.3.5.so
## 
## locale:
##  [1] LC_CTYPE=en_US.utf8       LC_NUMERIC=C             
##  [3] LC_TIME=en_US.utf8        LC_COLLATE=en_US.utf8    
##  [5] LC_MONETARY=en_US.utf8    LC_MESSAGES=en_US.utf8   
##  [7] LC_PAPER=en_US.utf8       LC_NAME=C                
##  [9] LC_ADDRESS=C              LC_TELEPHONE=C           
## [11] LC_MEASUREMENT=en_US.utf8 LC_IDENTIFICATION=C      
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] org.Cfasciculata.Cf.Cl.v41.eg.db_2019.02
##  [2] bindrcpp_0.2.2                          
##  [3] AnnotationDbi_1.44.0                    
##  [4] EuPathDB_1.0.1                          
##  [5] AnnotationHub_2.14.2                    
##  [6] GenomeInfoDbData_1.2.0                  
##  [7] GenomicRanges_1.34.0                    
##  [8] GenomeInfoDb_1.18.1                     
##  [9] IRanges_2.16.0                          
## [10] S4Vectors_0.20.1                        
## [11] Biobase_2.42.0                          
## [12] BiocGenerics_0.28.0                     
## [13] BiocStyle_2.10.0                        
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6                  matrixStats_0.54.0           
##  [3] fs_1.2.6                      usethis_1.4.0                
##  [5] devtools_2.0.1                bit64_0.9-7                  
##  [7] progress_1.2.0                httr_1.4.0                   
##  [9] rprojroot_1.3-2               tools_3.5.2                  
## [11] backports_1.1.3               utf8_1.1.4                   
## [13] R6_2.3.0                      DBI_1.0.0                    
## [15] withr_2.1.2                   tidyselect_0.2.5             
## [17] prettyunits_1.0.2             processx_3.2.1               
## [19] bit_1.1-14                    curl_3.3                     
## [21] compiler_3.5.2                cli_1.0.1                    
## [23] rvest_0.3.2                   xml2_1.2.0                   
## [25] desc_1.2.0                    DelayedArray_0.8.0           
## [27] rtracklayer_1.42.1            callr_3.1.1                  
## [29] stringr_1.3.1                 digest_0.6.18                
## [31] Rsamtools_1.34.0              rmarkdown_1.11               
## [33] XVector_0.22.0                pkgconfig_2.0.2              
## [35] htmltools_0.3.6               sessioninfo_1.1.1            
## [37] rlang_0.3.1                   RSQLite_2.1.1                
## [39] BiocInstaller_1.32.1          shiny_1.2.0                  
## [41] bindr_0.1.1                   jsonlite_1.6                 
## [43] BiocParallel_1.16.5           dplyr_0.7.8                  
## [45] RCurl_1.95-4.11               magrittr_1.5                 
## [47] GO.db_3.7.0                   Matrix_1.2-15                
## [49] fansi_0.4.0                   Rcpp_1.0.0                   
## [51] stringi_1.2.4                 yaml_2.2.0                   
## [53] SummarizedExperiment_1.12.0   zlibbioc_1.28.0              
## [55] pkgbuild_1.0.2                grid_3.5.2                   
## [57] blob_1.1.1                    promises_1.0.1               
## [59] crayon_1.3.4                  lattice_0.20-38              
## [61] Biostrings_2.50.2             GenomicFeatures_1.34.1       
## [63] hms_0.4.2                     knitr_1.21                   
## [65] ps_1.3.0                      pillar_1.3.1                 
## [67] codetools_0.2-16              biomaRt_2.38.0               
## [69] pkgload_1.0.2                 XML_3.98-1.16                
## [71] glue_1.3.0                    evaluate_0.12                
## [73] data.table_1.12.0             remotes_2.0.2                
## [75] BiocManager_1.30.4            httpuv_1.4.5.1               
## [77] foreach_1.4.4                 testthat_2.0.1               
## [79] purrr_0.2.5                   assertthat_0.2.0             
## [81] xfun_0.4                      mime_0.6                     
## [83] xtable_1.8-3                  later_0.7.5.9000             
## [85] tibble_2.0.1                  iterators_1.0.10             
## [87] GenomicAlignments_1.18.1      memoise_1.1.0                
## [89] interactiveDisplayBase_1.20.0</code></pre>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
