#' Create a series of POST requests which download all the annotation data for a species.
#'
#' The only way I have figured out how to download mass data from the eupathdb
#' is to ask for a raw dump of all available data using the GenesByGeneType
#' WADL.  Therefore, this function iterates over the various sequence types that
#' I have noticed at the eupathdb and does that for each type.
#'
#' @param entry Eupathdb annotation entry.
#' @param workdir Location to dump the resulting data.
#' @param overwrite Overwrite existing data if it exists?
post_eupath_annotations <- function(entry = NULL, workdir = "EuPathDB", overwrite = FALSE) {
    if (is.null(entry)) {
        stop("  Need an entry from the eupathdb.")
    }

    ## Check for output rda directory and create it if necessary
    rdadir <- file.path(workdir, "rda")
    if (!file.exists(rdadir)) {
        created <- dir.create(rdadir, recursive = TRUE)
    }

    ## Look for an existing savefile and load if we are not overwriting existing data.
    savefile <- file.path(rdadir, glue::glue("{entry[['Genome']]}_annotations.rda"))
    if (file.exists(savefile)) {
        if (isTRUE(overwrite)) {
            removed <- file.remove(savefile)
        } else {
            message("  Delete the file ", savefile, " to regenerate.")
            result <- new.env()
            load(savefile, envir = result)
            result <- result[["result"]]
            return(result)
        }
    }

    ## query body as a structured list
    ## This list was generated by going to:
    ## view-source:http://tritrypdb.org/webservices/GeneQuestions/GenesByMolecularWeight.wadl
    ## scrolling down to the 'o-fields' section, and writing down the most likely
    ## useful column names.
    ## I later came through and wrote a function function to automagically populate this list.
    species <- entry[["TaxonUnmodified"]]
    test_types <- get_eupath_gene_types()
    types <- c("protein coding", "snRNA encoding", "SRP RNA encoding",
               "tRNA encoding", "miRNA encoding", "snoRNA encoding",
               "rRNA encoding", "misc RNA", "scRNA encoding")
    result <- data.frame()
    for (t in 1:length(types)) {
        type <- types[t]
        message(" Querying gene type: ", type, ".")

        ## Set the parameters and gather the results.
        parameters <- list(
            "organism" = jsonlite::unbox(species),
            "geneType" = jsonlite::unbox(type),
            "includePseudogenes" = jsonlite::unbox("Yes"))
        question <- "GeneQuestions.GenesByGeneType"
        a_result <- try(post_eupath_raw(
            entry, question = question,
            parameters = parameters, table_name = "annot"), silent = TRUE)
        if ("try-error" %in% class(a_result)) {
            next
        } else {
            colnames(a_result) <- tolower(colnames(a_result))
            colnames(a_result) <- gsub(x = colnames(a_result),
                                       pattern = "annot_annotated",
                                       replacement = "annot")
            result <- rbind(result, a_result)
            message("Added ", nrow(a_result), " rows on ",
                    ncol(a_result), " columns to the data.")
        }
    }

    ## Get rid of spurious text in the previous IDs column
    if (!is.null(result[["annot_gene_previous_ids"]])) {
        result[["annot_gene_previous_ids"]] <- gsub(pattern = "^Previous IDs: ", replacement = "",
                                                    x = result[["annot_gene_previous_ids"]])
    }

    all_columns <- colnames(result)
    numeric_columns <- c(
        "annot_gene_exon_count",
        "annot_gene_transcript_count",
        "annot_gene_ortholog_number",
        "annot_gene_paralog_number",
        "annot_gene_total_hts_snps",
        "annot_gene_hts_nonsynonymous_snps",
        "annot_gene_hts_synonymous_snps",
        "annot_gene_hts_noncoding_snps",
        "annot_gene_hts_stop_codon_snps",
        "annot_gene_hts_nonsyn_syn_ratio",
        "annot_transcript_length",
        "annot_exon_count",
        "annot_cds_length",
        "annot_tm_count",
        "annot_molecular_weight",
        "annot_isoelectric_point",
        "annot_five_prime_utr_length",
        "annot_three_prime_utr_length")

    for (col in all_columns) {
        na_idx <- is.na(result[[col]])
        if (col %in% numeric_columns) {
            if (!is.null(result[[col]])) {
                result[[col]] <- as.numeric(result[[col]])
            }
            if (sum(na_idx) > 0) {
                result[na_idx, col] <- 0
            }
        } else {
            if (sum(na_idx) > 0) {
                result[na_idx, col] <- ""
            }
        }
    }

    ## Cast factor columns explicitly as factors
    factor_columns <- c(
        "annot_chromosome",
        "annot_gene_type",
        "annot_is_pseudo",
        "annot_strand",
        "annot_tm_count",
        "annot_exon_count")
    for (col in factor_columns) {
        if (!is.null(result[[col]])) {
            result[[col]] <- as.factor(result[[col]])
        }
        na_idx <- is.na(result[[col]])
        if (sum(na_idx) > 0) {
            result[na_idx, col] <- 0
        }
    }

    ## orgdbs seem to like uppercase column names
    colnames(result) <- toupper(colnames(result))

    message("  Saving ", savefile, " with ", nrow(result), " rows.")
    save(result, file = savefile)
    return(result)
}

#' Attempt to get a list of sequence types.
#'
#' @param webservice choose a service to download from.
get_eupath_gene_types <- function(webservice = NULL) {
    if (is.null(webservice)) {
        webservice <- "fungidb"
    }
    tld <- "org"
    if (webservice == "schistodb") {
        tld <- "net"
    }
    request_url <- glue::glue("http://{webservice}.{tld}/webservices/GeneQuestions/GenesByGeneType.wadl")
    request <- curl::curl(request_url)
    result <- xml2::read_xml(request)
    ##close(request)
    fields <- rvest::xml_nodes(result, xpath = '//*[@name="geneType"]')[[1]] %>%
        xml2::xml_children() %>%
        xml2::xml_attr("value")
    return(fields)
}
