post_eupath_annotations <- function(entry=NULL, workdir="EuPathDB", overwrite=FALSE) {
    if (is.null(entry)) {
        stop("  Need an entry from the eupathdb.")
    }
    rdadir <- file.path(workdir, "rda")
    if (!file.exists(rdadir)) {
        created <- dir.create(rdadir, recursive=TRUE)
    }
    savefile <- file.path(rdadir, glue::glue("{entry[['Genome']]}_annotations.rda"))
    if (file.exists(savefile)) {
        if (isTRUE(overwrite)) {
            removed <- file.remove(savefile)
        } else {
            message("  Delete the file ", savefile, " to regenerate.")
            result <- new.env()
            load(savefile, envir=result)
            result <- result[["result"]]
            return(result)
        }
    }

    ## query body as a structured list
    ## This list was generated by going to:
    ## view-source:http://tritrypdb.org/webservices/GeneQuestions/GenesByMolecularWeight.wadl
    ## scrolling down to the 'o-fields' section, and writing down the most likely
    ## useful column names. These are written one per line in an attempt to make
    ## looking for new/changed columns from one eupathdb release to the next
    ## easier.
    species <- entry[["TaxonUnmodified"]]
    test_types <- get_eupath_gene_types()
    ## In theory, my little get_eupath_gene_types() should get the below
    ## But I am not ready to assume that, given that xml2 and httr are confusing to me.
    types <- c("protein coding", "snRNA encoding", "SRP RNA encoding",
               "tRNA encoding", "miRNA encoding", "snoRNA encoding",
               "rRNA encoding", "misc RNA", "scRNA encoding")
    result <- data.frame()
    for (t in 1:length(types)) {
        type <- types[t]
        message(" Querying gene type: ", type, ".")
        parameters <- list(
            "organism" = jsonlite::unbox(species),
            "geneType" = jsonlite::unbox(type),
            "includePseudogenes" = jsonlite::unbox("Yes"))
        question <- "GeneQuestions.GenesByGeneType"
        a_result <- try(post_eupath_raw(entry, question=question,
                                        parameters=parameters, table_name="annot"), silent=TRUE)
        if ("try-error" %in% class(a_result)) {
            next
        } else {
            colnames(a_result) <- tolower(colnames(a_result))
            colnames(a_result) <- gsub(x=colnames(a_result),
                                       pattern="annot_annotated",
                                       replacement="annot")
            result <- rbind(result, a_result)
            message("Added ", nrow(a_result), " rows on ",
                    ncol(a_result), " columns to the data.")
        }
    }
    ## Get rid of spurious text in the previous IDs column
    if (!is.null(result[["annot_gene_previous_ids"]])) {
        result[["annot_gene_previous_ids"]] <- gsub(pattern="^Previous IDs: ", replacement="",
                                                    x=result[["annot_gene_previous_ids"]])
    }
    all_columns <- colnames(result)
    numeric_columns <- c(
        "annot_gene_exon_count",
        "annot_gene_transcript_count",
        "annot_gene_ortholog_number",
        "annot_gene_paralog_number",
        "annot_gene_total_hts_snps",
        "annot_gene_hts_nonsynonymous_snps",
        "annot_gene_hts_synonymous_snps",
        "annot_gene_hts_noncoding_snps",
        "annot_gene_hts_stop_codon_snps",
        "annot_gene_hts_nonsyn_syn_ratio",
        "annot_transcript_length",
        "annot_exon_count",
        "annot_cds_length",
        "annot_tm_count",
        "annot_molecular_weight",
        "annot_isoelectric_point",
        "annot_five_prime_utr_length",
        "annot_three_prime_utr_length")
    for (col in all_columns) {
        na_idx <- is.na(result[[col]])
        if (col %in% numeric_columns) {
            if (!is.null(result[[col]])) {
                result[[col]] <- as.numeric(result[[col]])
            }
            if (sum(na_idx) > 0) {
                result[na_idx, col] <- 0
            }
        } else {
            if (sum(na_idx) > 0) {
                result[na_idx, col] <- ""
            }
        }
    }
    factor_columns <- c(
        "annot_chromosome",
        "annot_gene_type",
        "annot_is_pseudo",
        "annot_strand",
        "annot_tm_count",
        "annot_exon_count")
    for (col in factor_columns) {
        if (!is.null(result[[col]])) {
            result[[col]] <- as.factor(result[[col]])
        }
        na_idx <- is.na(result[[col]])
        if (sum(na_idx) > 0) {
            result[na_idx, col] <- 0
        }
    }
    colnames(result) <- toupper(colnames(result))

    message("  Saving ", savefile)
    save(result, file=savefile)
    return(result)
}

get_eupath_gene_types <- function(webservice=NULL) {
    if (is.null(webservice)) {
        webservice <- "fungidb"
    }
    tld <- "org"
    if (webservice == "schistodb") {
        tld <- "net"
    }
    request_url <- glue::glue(
                             "http://{webservice}.{tld}/webservices/GeneQuestions/GenesByGeneType.wadl")
    request <- curl::curl(request_url)
    result <- xml2::read_xml(request)
    ##close(request)
    fields <- rvest::xml_nodes(result, xpath='//*[@name="geneType"]')[[1]] %>%
        xml2::xml_children() %>%
        xml2::xml_attr("value")
    return(fields)
}

#' Gather all available annotation data for a given eupathdb species.
#'
#' This function fills in the parameters to post_eupath_raw() so that one can
#' download all the available data for a given parasite into one massive table.
#' It should also provide some constraints to the data rather than leaving it
#' all as characters.  Caveat:  I manually filled in the list 'field_list' to
#' include the variable names and their text associations.  This is likely to
#' change in future releases of the tritrypdb.  It is probably possible to
#' automagically fill it in.  In addition, I am using GenesByMolecularWeight to
#' get the data, which is a bit weird.
#'
#' @param entry The full annotation entry.
#' @param workdir A directory into which to write the intermediate savefile.
#' @param overwrite If a partial table exists, overwrite it?
#' @return  A big honking table.
old_post_eupath_annotations <- function(entry=NULL, workdir="EuPathDB", overwrite=FALSE) {
    if (is.null(entry)) {
        stop("  Need an entry from the eupathdb.")
    }
    rdadir <- file.path(workdir, "rda")
    if (!file.exists(rdadir)) {
        created <- dir.create(rdadir, recursive=TRUE)
    }
    savefile <- file.path(rdadir, glue::glue("{entry[['Genome']]}_annotations.rda"))
    if (file.exists(savefile)) {
        if (isTRUE(overwrite)) {
            removed <- file.remove(savefile)
        } else {
            message("  Delete the file ", savefile, " to regenerate.")
            result <- new.env()
            load(savefile, envir=result)
            result <- result[["result"]]
            return(result)
        }
    }

    ## query body as a structured list
    ## This list was generated by going to:
    ## view-source:http://tritrypdb.org/webservices/GeneQuestions/GenesByMolecularWeight.wadl
    ## scrolling down to the 'o-fields' section, and writing down the most likely
    ## useful column names. These are written one per line in an attempt to make
    ## looking for new/changed columns from one eupathdb release to the next
    ## easier.
    species <- entry[["TaxonUnmodified"]]
    parameters <- list(
        "organism" = jsonlite::unbox(species),
        "min_molecular_weight" = jsonlite::unbox("0"),
        "max_molecular_weight" = jsonlite::unbox("10000000000000000")
    )
    question <- "GeneQuestions.GenesByMolecularWeight"
    result <- post_eupath_raw(entry, question=question,
                              parameters=parameters, table_name="annot")
    colnames(result) <- tolower(colnames(result))
    colnames(result) <- gsub(x=colnames(result), pattern="annot_annotated", replacement="annot")
    ## Get rid of spurious text in the previous IDs column
    if (!is.null(result[["annot_gene_previous_ids"]])) {
        result[["annot_gene_previous_ids"]] <- gsub(pattern="^Previous IDs: ", replacement="",
                                                    x=result[["annot_gene_previous_ids"]])
    }
    all_columns <- colnames(result)
    numeric_columns <- c(
        "annot_gene_exon_count",
        "annot_gene_transcript_count",
        "annot_gene_ortholog_number",
        "annot_gene_paralog_number",
        "annot_gene_total_hts_snps",
        "annot_gene_hts_nonsynonymous_snps",
        "annot_gene_hts_synonymous_snps",
        "annot_gene_hts_noncoding_snps",
        "annot_gene_hts_stop_codon_snps",
        "annot_gene_hts_nonsyn_syn_ratio",
        "annot_transcript_length",
        "annot_exon_count",
        "annot_cds_length",
        "annot_tm_count",
        "annot_molecular_weight",
        "annot_isoelectric_point",
        "annot_five_prime_utr_length",
        "annot_three_prime_utr_length")
    for (col in all_columns) {
        na_idx <- is.na(result[[col]])
        if (col %in% numeric_columns) {
            if (!is.null(result[[col]])) {
                result[[col]] <- as.numeric(result[[col]])
            }
            if (sum(na_idx) > 0) {
                result[na_idx, col] <- 0
            }
        } else {
            if (sum(na_idx) > 0) {
                result[na_idx, col] <- ""
            }
        }
    }
    factor_columns <- c(
        "annot_chromosome",
        "annot_gene_type",
        "annot_is_pseudo",
        "annot_strand",
        "annot_tm_count",
        "annot_exon_count")
    for (col in factor_columns) {
        if (!is.null(result[[col]])) {
            result[[col]] <- as.factor(result[[col]])
        }
        na_idx <- is.na(result[[col]])
        if (sum(na_idx) > 0) {
            result[na_idx, col] <- 0
        }
    }
    colnames(result) <- toupper(colnames(result))

    message("  Saving ", savefile)
    save(result, file=savefile)
    return(result)
}
